{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "d1d35981-a41d-42a4-8e7a-8cbf0b3292e8",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "library(Signac)\n",
    "library(Seurat)\n",
    "library(EnsDb.Hsapiens.v86)\n",
    "library(SeuratDisk)\n",
    "library(dplyr)\n",
    "library(BSgenome.Hsapiens.UCSC.hg38)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d1f66ad8-d022-438a-8ad6-eb1969912570",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "ifnb <- readRDS(\"pd_processed_rna+atac.rds\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "4ea60d68-7f86-4437-bb6b-ddd6da615eb0",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "# ifnb <- subset(ifnb, celltype == \"microglia\")\n",
    "# ifnb <- subset(ifnb, celltype == \"excitatory neurons\")\n",
    "ifnb <- subset(ifnb, celltype == \"inhibitory neurons\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "2a1ebeb0-2405-48f7-90b4-e34dd927787d",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'chrX'</li><li>'chr20'</li><li>'chr1'</li><li>'chr6'</li><li>'chr3'</li><li>'chr7'</li><li>'chr12'</li><li>'chr11'</li><li>'chr4'</li><li>'chr17'</li><li>'chr2'</li><li>'chr16'</li><li>'chr8'</li><li>'chr19'</li><li>'chr9'</li><li>'chr13'</li><li>'chr14'</li><li>'chr5'</li><li>'chr22'</li><li>'chr10'</li><li>'chrY'</li><li>'chr18'</li><li>'chr15'</li><li>'chr21'</li><li>'chrMT'</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 'chrX'\n",
       "\\item 'chr20'\n",
       "\\item 'chr1'\n",
       "\\item 'chr6'\n",
       "\\item 'chr3'\n",
       "\\item 'chr7'\n",
       "\\item 'chr12'\n",
       "\\item 'chr11'\n",
       "\\item 'chr4'\n",
       "\\item 'chr17'\n",
       "\\item 'chr2'\n",
       "\\item 'chr16'\n",
       "\\item 'chr8'\n",
       "\\item 'chr19'\n",
       "\\item 'chr9'\n",
       "\\item 'chr13'\n",
       "\\item 'chr14'\n",
       "\\item 'chr5'\n",
       "\\item 'chr22'\n",
       "\\item 'chr10'\n",
       "\\item 'chrY'\n",
       "\\item 'chr18'\n",
       "\\item 'chr15'\n",
       "\\item 'chr21'\n",
       "\\item 'chrMT'\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 'chrX'\n",
       "2. 'chr20'\n",
       "3. 'chr1'\n",
       "4. 'chr6'\n",
       "5. 'chr3'\n",
       "6. 'chr7'\n",
       "7. 'chr12'\n",
       "8. 'chr11'\n",
       "9. 'chr4'\n",
       "10. 'chr17'\n",
       "11. 'chr2'\n",
       "12. 'chr16'\n",
       "13. 'chr8'\n",
       "14. 'chr19'\n",
       "15. 'chr9'\n",
       "16. 'chr13'\n",
       "17. 'chr14'\n",
       "18. 'chr5'\n",
       "19. 'chr22'\n",
       "20. 'chr10'\n",
       "21. 'chrY'\n",
       "22. 'chr18'\n",
       "23. 'chr15'\n",
       "24. 'chr21'\n",
       "25. 'chrMT'\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       " [1] \"chrX\"  \"chr20\" \"chr1\"  \"chr6\"  \"chr3\"  \"chr7\"  \"chr12\" \"chr11\" \"chr4\" \n",
       "[10] \"chr17\" \"chr2\"  \"chr16\" \"chr8\"  \"chr19\" \"chr9\"  \"chr13\" \"chr14\" \"chr5\" \n",
       "[19] \"chr22\" \"chr10\" \"chrY\"  \"chr18\" \"chr15\" \"chr21\" \"chrMT\""
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "seqlevels(Annotation(ifnb))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "83524526-3f1f-4a43-aa1f-7fe619c1e3ad",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in RegionStats.default(object = regions, genome = genome, verbose = verbose, :\n",
      "\"Not all seqlevels present in supplied genome\"\n"
     ]
    }
   ],
   "source": [
    "DefaultAssay(ifnb) <- \"ATAC\"\n",
    "# first compute the GC content for each peak\n",
    "ifnb <- RegionStats(ifnb, genome = BSgenome.Hsapiens.UCSC.hg38)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "84a22762-4062-4511-9065-373d5008be51",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "# idents.plot <- c('microglia')\n",
    "# idents.plot <- c('excitatory neurons')\n",
    "idents.plot <- c('inhibitory neurons')\n",
    "Idents(ifnb) <- idents.plot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "e8673cad-6857-45e9-adfa-e41dd77fc12e",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "An object of class Seurat \n",
       "283535 features across 3667 samples within 3 assays \n",
       "Active assay: ATAC (216774 features, 207802 variable features)\n",
       " 2 layers present: counts, data\n",
       " 2 other assays present: RNA, SCT\n",
       " 2 dimensional reductions calculated: pca, lsi"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "ifnb"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "af371614-cbc0-4966-8344-d4a847ca54fe",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Testing 1540 genes and 116887 peaks\n",
      "\n",
      "Found gene coordinates for 1102 genes\n",
      "\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"Each of the 2 combined objects has sequence levels not in the other:\n",
      "  - in 'x': GL000009.2, GL000194.1, GL000195.1, GL000205.2, GL000218.1, GL000219.1, KI270711.1, KI270713.1, KI270727.1, KI270728.1, KI270731.1, KI270734.1, KI270721.1, KI270726.1\n",
      "  - in 'y': chrMT\n",
      "  Make sure to always combine/compare objects based on the same reference\n",
      "  genome (use suppressWarnings() to suppress this warning).\"\n",
      "Warning message:\n",
      "\"package 'tidyverse' was built under R version 4.3.3\"\n",
      "Warning message:\n",
      "\"package 'ggplot2' was built under R version 4.3.3\"\n",
      "Warning message:\n",
      "\"package 'tibble' was built under R version 4.3.3\"\n",
      "Warning message:\n",
      "\"package 'tidyr' was built under R version 4.3.3\"\n",
      "Warning message:\n",
      "\"package 'readr' was built under R version 4.3.3\"\n",
      "Warning message:\n",
      "\"package 'purrr' was built under R version 4.3.3\"\n",
      "Warning message:\n",
      "\"package 'stringr' was built under R version 4.3.3\"\n",
      "Warning message:\n",
      "\"package 'forcats' was built under R version 4.3.3\"\n",
      "Warning message:\n",
      "\"package 'lubridate' was built under R version 4.3.3\"\n",
      "── \u001b[1mAttaching core tidyverse packages\u001b[22m ──────────────────────── tidyverse 2.0.0 ──\n",
      "\u001b[32m✔\u001b[39m \u001b[34mforcats  \u001b[39m 1.0.0     \u001b[32m✔\u001b[39m \u001b[34mreadr    \u001b[39m 2.1.5\n",
      "\u001b[32m✔\u001b[39m \u001b[34mggplot2  \u001b[39m 3.5.1     \u001b[32m✔\u001b[39m \u001b[34mstringr  \u001b[39m 1.5.1\n",
      "\u001b[32m✔\u001b[39m \u001b[34mlubridate\u001b[39m 1.9.3     \u001b[32m✔\u001b[39m \u001b[34mtibble   \u001b[39m 3.2.1\n",
      "\u001b[32m✔\u001b[39m \u001b[34mpurrr    \u001b[39m 1.0.2     \u001b[32m✔\u001b[39m \u001b[34mtidyr    \u001b[39m 1.3.1\n",
      "── \u001b[1mConflicts\u001b[22m ────────────────────────────────────────── tidyverse_conflicts() ──\n",
      "\u001b[31m✖\u001b[39m \u001b[34mlubridate\u001b[39m::\u001b[32m%within%()\u001b[39m  masks \u001b[34mIRanges\u001b[39m::%within%()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mBiostrings\u001b[39m::\u001b[32mcollapse()\u001b[39m masks \u001b[34mdplyr\u001b[39m::collapse(), \u001b[34mIRanges\u001b[39m::collapse()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mcombine()\u001b[39m       masks \u001b[34mBiobase\u001b[39m::combine(), \u001b[34mBiocGenerics\u001b[39m::combine()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mpurrr\u001b[39m::\u001b[32mcompact()\u001b[39m       masks \u001b[34mXVector\u001b[39m::compact()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mdesc()\u001b[39m          masks \u001b[34mIRanges\u001b[39m::desc()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mtidyr\u001b[39m::\u001b[32mexpand()\u001b[39m        masks \u001b[34mS4Vectors\u001b[39m::expand()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mfilter()\u001b[39m        masks \u001b[34mensembldb\u001b[39m::filter(), \u001b[34mstats\u001b[39m::filter()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mfirst()\u001b[39m         masks \u001b[34mS4Vectors\u001b[39m::first()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mlag()\u001b[39m           masks \u001b[34mstats\u001b[39m::lag()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mggplot2\u001b[39m::\u001b[32mPosition()\u001b[39m    masks \u001b[34mBiocGenerics\u001b[39m::Position(), \u001b[34mbase\u001b[39m::Position()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mpurrr\u001b[39m::\u001b[32mreduce()\u001b[39m        masks \u001b[34mGenomicRanges\u001b[39m::reduce(), \u001b[34mIRanges\u001b[39m::reduce(), \u001b[34mSignac\u001b[39m::reduce()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mrename()\u001b[39m        masks \u001b[34mS4Vectors\u001b[39m::rename()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mlubridate\u001b[39m::\u001b[32msecond()\u001b[39m    masks \u001b[34mS4Vectors\u001b[39m::second()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mlubridate\u001b[39m::\u001b[32msecond<-()\u001b[39m  masks \u001b[34mS4Vectors\u001b[39m::second<-()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mselect()\u001b[39m        masks \u001b[34mensembldb\u001b[39m::select(), \u001b[34mAnnotationDbi\u001b[39m::select()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mXVector\u001b[39m::\u001b[32mslice()\u001b[39m       masks \u001b[34mdplyr\u001b[39m::slice(), \u001b[34mIRanges\u001b[39m::slice()\n",
      "\u001b[36mℹ\u001b[39m Use the conflicted package (\u001b[3m\u001b[34m<http://conflicted.r-lib.org/>\u001b[39m\u001b[23m) to force all conflicts to become errors\n",
      "Saving data from SCT as /matrix\n",
      "\n",
      "Adding slot counts for assay SCT\n",
      "\n",
      "Adding layer counts\n",
      "\n",
      "Adding col attribute CellID\n",
      "\n",
      "Adding col attribute orig.ident\n",
      "\n",
      "Adding col attribute nCount_RNA\n",
      "\n",
      "Adding col attribute nFeature_RNA\n",
      "\n",
      "Adding col attribute nCount_ATAC\n",
      "\n",
      "Adding col attribute nFeature_ATAC\n",
      "\n",
      "Adding col attribute nucleosome_signal\n",
      "\n",
      "Adding col attribute nucleosome_percentile\n",
      "\n",
      "Adding col attribute TSS.enrichment\n",
      "\n",
      "Adding col attribute TSS.percentile\n",
      "\n",
      "Adding col attribute celltype\n",
      "\n",
      "Adding col attribute batch\n",
      "\n",
      "Adding col attribute nCount_SCT\n",
      "\n",
      "Adding col attribute nFeature_SCT\n",
      "\n",
      "Adding row attribute Gene\n",
      "\n"
     ]
    }
   ],
   "source": [
    "total_genes <- c()\n",
    "library(readxl)\n",
    "library(dplyr)\n",
    "\n",
    "# total_genes <- readLines(\"../../Signac_plots/saved_files/mic_total_genes.txt\")\n",
    "# total_genes <- readLines(\"../../Signac_plots/saved_files/exn_rorb_total_genes.txt\")\n",
    "total_genes <- readLines(\"../../Signac_plots/saved_files/inh_subset_total_genes.txt\")\n",
    "\n",
    "\n",
    "# 1.2 Pull out the links that LinkPeaks() created\n",
    "## each element of the GRanges object is one peak; the linked gene lives in metadata\n",
    "ifnb <- LinkPeaks(\n",
    "  object = ifnb,\n",
    "  peak.assay = \"ATAC\",\n",
    "  expression.assay = \"SCT\",\n",
    "  genes.use = total_genes\n",
    ")\n",
    "\n",
    "peak_gene_links <- Links(ifnb[[\"ATAC\"]])\n",
    "df <- as.data.frame(peak_gene_links)\n",
    "\n",
    "links_df <- as_tibble(df) |>\n",
    "  transmute(\n",
    "    chrom      = seqnames,          # chromosome\n",
    "    chromStart = start              # 0-based for BED\n",
    "      - 1L,                         # Signac stores 1-based; BED wants 0-based half-open\n",
    "    chromEnd   = end,\n",
    "    gene       = gene,\n",
    "    score      = score              # optional\n",
    "  )\n",
    "\n",
    "links_df <- links_df |>\n",
    "  mutate(peak_id = paste0(chrom, \":\", chromStart, \"-\", chromEnd))\n",
    "\n",
    "library(tidyverse)\n",
    "# write_tsv(links_df,  \"../../Signac_plots/saved_files/PD_gene_peak_links_mic.tsv\")\n",
    "# write_tsv(links_df,  \"../../Signac_plots/saved_files/PD_gene_peak_links_exn_rorb.tsv\")\n",
    "write_tsv(links_df,  \"../../Signac_plots/saved_files/PD_gene_peak_links_inh_subset.tsv\")\n",
    "\n",
    "\n",
    "peaks_df <- links_df %>% \n",
    "  select(chrom, chromStart, chromEnd, peak_id)\n",
    "\n",
    "# readr::write_tsv(\n",
    "#   peaks_df, \n",
    "#   \"../../Signac_plots/saved_files/PD_peaks_mic.bed\", \n",
    "#   col_names = FALSE\n",
    "# )\n",
    "# readr::write_tsv(\n",
    "#   peaks_df, \n",
    "#   \"../../Signac_plots/saved_files/PD_peaks_exn_rorb.bed\", \n",
    "#   col_names = FALSE\n",
    "# )\n",
    "readr::write_tsv(\n",
    "  peaks_df, \n",
    "  \"../../Signac_plots/saved_files/PD_peaks_inh_subset.bed\", \n",
    "  col_names = FALSE\n",
    ")\n",
    "\n",
    "tfs <- read.table(\"../../Signac_plots/saved_files/all_tfs.txt\", stringsAsFactors = FALSE)[,1]\n",
    "\n",
    "DefaultAssay(ifnb) <- \"SCT\"\n",
    "keep_genes <- union(VariableFeatures(ifnb), tfs)\n",
    "ifnb_subset <- subset(ifnb, features = keep_genes)\n",
    "\n",
    "# SaveLoom(\n",
    "#   object   = ifnb_subset,\n",
    "#   filename = \"../../Signac_plots/saved_files/pd_mic_all_tfs_rna_MIC.loom\",\n",
    "#   assay    = \"SCT\",\n",
    "#   slot     = \"data\"\n",
    "# )\n",
    "# SaveLoom(\n",
    "#   object   = ifnb_subset,\n",
    "#   filename = \"../../Signac_plots/saved_files/pd_exn_all_tfs_rna_EXN_RORB.loom\",\n",
    "#   assay    = \"SCT\",\n",
    "#   slot     = \"data\"\n",
    "# )\n",
    "SaveLoom(\n",
    "  object   = ifnb_subset,\n",
    "  filename = \"../../Signac_plots/saved_files/pd_inh_all_tfs_rna_INH_SUBSET.loom\",\n",
    "  assay    = \"SCT\",\n",
    "  slot     = \"data\"\n",
    ")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.3.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
