{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "d1d35981-a41d-42a4-8e7a-8cbf0b3292e8",
   "metadata": {
    "scrolled": true,
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "“package ‘Signac’ was built under R version 4.3.3”\n",
      "Warning message:\n",
      "“package ‘Seurat’ was built under R version 4.3.3”\n",
      "Loading required package: SeuratObject\n",
      "\n",
      "Warning message:\n",
      "“package ‘SeuratObject’ was built under R version 4.3.3”\n",
      "Loading required package: sp\n",
      "\n",
      "Warning message:\n",
      "“package ‘sp’ was built under R version 4.3.3”\n",
      "\n",
      "Attaching package: ‘SeuratObject’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    intersect, t\n",
      "\n",
      "\n",
      "Loading required package: ensembldb\n",
      "\n",
      "Warning message:\n",
      "“package ‘ensembldb’ was built under R version 4.3.2”\n",
      "Loading required package: BiocGenerics\n",
      "\n",
      "Warning message:\n",
      "“package ‘BiocGenerics’ was built under R version 4.3.2”\n",
      "\n",
      "Attaching package: ‘BiocGenerics’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:SeuratObject’:\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    IQR, mad, sd, var, xtabs\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    anyDuplicated, aperm, append, as.data.frame, basename, cbind,\n",
      "    colnames, dirname, do.call, duplicated, eval, evalq, Filter, Find,\n",
      "    get, grep, grepl, intersect, is.unsorted, lapply, Map, mapply,\n",
      "    match, mget, order, paste, pmax, pmax.int, pmin, pmin.int,\n",
      "    Position, rank, rbind, Reduce, rownames, sapply, setdiff, sort,\n",
      "    table, tapply, union, unique, unsplit, which.max, which.min\n",
      "\n",
      "\n",
      "Loading required package: GenomicRanges\n",
      "\n",
      "Warning message:\n",
      "“package ‘GenomicRanges’ was built under R version 4.3.3”\n",
      "Loading required package: stats4\n",
      "\n",
      "Loading required package: S4Vectors\n",
      "\n",
      "Warning message:\n",
      "“package ‘S4Vectors’ was built under R version 4.3.3”\n",
      "\n",
      "Attaching package: ‘S4Vectors’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:utils’:\n",
      "\n",
      "    findMatches\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    expand.grid, I, unname\n",
      "\n",
      "\n",
      "Loading required package: IRanges\n",
      "\n",
      "Warning message:\n",
      "“package ‘IRanges’ was built under R version 4.3.3”\n",
      "\n",
      "Attaching package: ‘IRanges’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:sp’:\n",
      "\n",
      "    %over%\n",
      "\n",
      "\n",
      "Loading required package: GenomeInfoDb\n",
      "\n",
      "Warning message:\n",
      "“package ‘GenomeInfoDb’ was built under R version 4.3.2”\n",
      "Loading required package: GenomicFeatures\n",
      "\n",
      "Warning message:\n",
      "“package ‘GenomicFeatures’ was built under R version 4.3.2”\n",
      "Loading required package: AnnotationDbi\n",
      "\n",
      "Warning message:\n",
      "“package ‘AnnotationDbi’ was built under R version 4.3.2”\n",
      "Loading required package: Biobase\n",
      "\n",
      "Warning message:\n",
      "“package ‘Biobase’ was built under R version 4.3.3”\n",
      "Welcome to Bioconductor\n",
      "\n",
      "    Vignettes contain introductory material; view with\n",
      "    'browseVignettes()'. To cite Bioconductor, see\n",
      "    'citation(\"Biobase\")', and for packages 'citation(\"pkgname\")'.\n",
      "\n",
      "\n",
      "Loading required package: AnnotationFilter\n",
      "\n",
      "Warning message:\n",
      "“package ‘AnnotationFilter’ was built under R version 4.3.2”\n",
      "\n",
      "Attaching package: 'ensembldb'\n",
      "\n",
      "\n",
      "The following object is masked from 'package:stats':\n",
      "\n",
      "    filter\n",
      "\n",
      "\n",
      "Registered S3 method overwritten by 'SeuratDisk':\n",
      "  method            from  \n",
      "  as.sparse.H5Group Seurat\n",
      "\n",
      "Warning message:\n",
      "\"package 'dplyr' was built under R version 4.3.3\"\n",
      "\n",
      "Attaching package: 'dplyr'\n",
      "\n",
      "\n",
      "The following objects are masked from 'package:ensembldb':\n",
      "\n",
      "    filter, select\n",
      "\n",
      "\n",
      "The following object is masked from 'package:AnnotationDbi':\n",
      "\n",
      "    select\n",
      "\n",
      "\n",
      "The following object is masked from 'package:Biobase':\n",
      "\n",
      "    combine\n",
      "\n",
      "\n",
      "The following objects are masked from 'package:GenomicRanges':\n",
      "\n",
      "    intersect, setdiff, union\n",
      "\n",
      "\n",
      "The following object is masked from 'package:GenomeInfoDb':\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "The following objects are masked from 'package:IRanges':\n",
      "\n",
      "    collapse, desc, intersect, setdiff, slice, union\n",
      "\n",
      "\n",
      "The following objects are masked from 'package:S4Vectors':\n",
      "\n",
      "    first, intersect, rename, setdiff, setequal, union\n",
      "\n",
      "\n",
      "The following objects are masked from 'package:BiocGenerics':\n",
      "\n",
      "    combine, intersect, setdiff, union\n",
      "\n",
      "\n",
      "The following objects are masked from 'package:stats':\n",
      "\n",
      "    filter, lag\n",
      "\n",
      "\n",
      "The following objects are masked from 'package:base':\n",
      "\n",
      "    intersect, setdiff, setequal, union\n",
      "\n",
      "\n",
      "Loading required package: BSgenome\n",
      "\n",
      "Warning message:\n",
      "\"package 'BSgenome' was built under R version 4.3.2\"\n",
      "Loading required package: Biostrings\n",
      "\n",
      "Warning message:\n",
      "\"package 'Biostrings' was built under R version 4.3.3\"\n",
      "Loading required package: XVector\n",
      "\n",
      "Warning message:\n",
      "\"package 'XVector' was built under R version 4.3.3\"\n",
      "\n",
      "Attaching package: 'Biostrings'\n",
      "\n",
      "\n",
      "The following object is masked from 'package:base':\n",
      "\n",
      "    strsplit\n",
      "\n",
      "\n",
      "Loading required package: BiocIO\n",
      "\n",
      "Warning message:\n",
      "\"package 'BiocIO' was built under R version 4.3.2\"\n",
      "Loading required package: rtracklayer\n",
      "\n",
      "Warning message:\n",
      "\"package 'rtracklayer' was built under R version 4.3.3\"\n",
      "\n",
      "Attaching package: 'rtracklayer'\n",
      "\n",
      "\n",
      "The following object is masked from 'package:BiocIO':\n",
      "\n",
      "    FileForFormat\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "library(Signac)\n",
    "library(Seurat)\n",
    "library(EnsDb.Hsapiens.v86)\n",
    "library(SeuratDisk)\n",
    "library(dplyr)\n",
    "library(BSgenome.Hsapiens.UCSC.hg38)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3ce866fd",
   "metadata": {},
   "source": [
    "To store RNA expression for relevant genes in each interested celltype (Mic, Exc, Inh), repeat this notebook for each cell type. The current version shows an example for Inh. If you wanted to do the same for other cell types, uncomment out filtering and saving lines for the relevant celltype."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "965ac80f-8777-40f1-aca5-b3f61ee01f61",
   "metadata": {},
   "source": [
    "# Get Peak - Gene Links"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c5ca525d",
   "metadata": {},
   "source": [
    "We only store expression for relevant genes for HC samples in this script."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d1f66ad8-d022-438a-8ad6-eb1969912570",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "ifnb <- readRDS(\"hc_processed_rna+atac.rds\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2dbdf5ba-8976-4273-acce-bfe94282e7bf",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "ifnb@meta.data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0ed19243-41c1-4759-8bb3-8a017137177f",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "# do this for each of the 3 cell types\n",
    "# ifnb <- subset(ifnb, celltype == \"microglia\")\n",
    "# ifnb <- subset(ifnb, celltype == \"excitatory neurons\")\n",
    "ifnb <- subset(ifnb, celltype == \"inhibitory neurons\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "2a1ebeb0-2405-48f7-90b4-e34dd927787d",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'chrX'</li><li>'chr20'</li><li>'chr1'</li><li>'chr6'</li><li>'chr3'</li><li>'chr7'</li><li>'chr12'</li><li>'chr11'</li><li>'chr4'</li><li>'chr17'</li><li>'chr2'</li><li>'chr16'</li><li>'chr8'</li><li>'chr19'</li><li>'chr9'</li><li>'chr13'</li><li>'chr14'</li><li>'chr5'</li><li>'chr22'</li><li>'chr10'</li><li>'chrY'</li><li>'chr18'</li><li>'chr15'</li><li>'chr21'</li><li>'chrMT'</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 'chrX'\n",
       "\\item 'chr20'\n",
       "\\item 'chr1'\n",
       "\\item 'chr6'\n",
       "\\item 'chr3'\n",
       "\\item 'chr7'\n",
       "\\item 'chr12'\n",
       "\\item 'chr11'\n",
       "\\item 'chr4'\n",
       "\\item 'chr17'\n",
       "\\item 'chr2'\n",
       "\\item 'chr16'\n",
       "\\item 'chr8'\n",
       "\\item 'chr19'\n",
       "\\item 'chr9'\n",
       "\\item 'chr13'\n",
       "\\item 'chr14'\n",
       "\\item 'chr5'\n",
       "\\item 'chr22'\n",
       "\\item 'chr10'\n",
       "\\item 'chrY'\n",
       "\\item 'chr18'\n",
       "\\item 'chr15'\n",
       "\\item 'chr21'\n",
       "\\item 'chrMT'\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 'chrX'\n",
       "2. 'chr20'\n",
       "3. 'chr1'\n",
       "4. 'chr6'\n",
       "5. 'chr3'\n",
       "6. 'chr7'\n",
       "7. 'chr12'\n",
       "8. 'chr11'\n",
       "9. 'chr4'\n",
       "10. 'chr17'\n",
       "11. 'chr2'\n",
       "12. 'chr16'\n",
       "13. 'chr8'\n",
       "14. 'chr19'\n",
       "15. 'chr9'\n",
       "16. 'chr13'\n",
       "17. 'chr14'\n",
       "18. 'chr5'\n",
       "19. 'chr22'\n",
       "20. 'chr10'\n",
       "21. 'chrY'\n",
       "22. 'chr18'\n",
       "23. 'chr15'\n",
       "24. 'chr21'\n",
       "25. 'chrMT'\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       " [1] \"chrX\"  \"chr20\" \"chr1\"  \"chr6\"  \"chr3\"  \"chr7\"  \"chr12\" \"chr11\" \"chr4\" \n",
       "[10] \"chr17\" \"chr2\"  \"chr16\" \"chr8\"  \"chr19\" \"chr9\"  \"chr13\" \"chr14\" \"chr5\" \n",
       "[19] \"chr22\" \"chr10\" \"chrY\"  \"chr18\" \"chr15\" \"chr21\" \"chrMT\""
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "seqlevels(Annotation(ifnb))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "83524526-3f1f-4a43-aa1f-7fe619c1e3ad",
   "metadata": {
    "scrolled": true,
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in RegionStats.default(object = regions, genome = genome, verbose = verbose, :\n",
      "\"Not all seqlevels present in supplied genome\"\n"
     ]
    }
   ],
   "source": [
    "DefaultAssay(ifnb) <- \"ATAC\"\n",
    "# first compute the GC content for each peak\n",
    "ifnb <- RegionStats(ifnb, genome = BSgenome.Hsapiens.UCSC.hg38)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "84a22762-4062-4511-9065-373d5008be51",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "# idents.plot <- c('microglia')\n",
    "# idents.plot <- c('excitatory neurons')\n",
    "idents.plot <- c('inhibitory neurons')\n",
    "Idents(ifnb) <- idents.plot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "36b5edfe-4200-4d55-813d-51de12d649ca",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "An object of class Seurat \n",
       "333440 features across 7739 samples within 3 assays \n",
       "Active assay: ATAC (265980 features, 247295 variable features)\n",
       " 2 layers present: counts, data\n",
       " 2 other assays present: RNA, SCT\n",
       " 2 dimensional reductions calculated: pca, lsi"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "ifnb"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cf44102d-af7f-4c31-be3e-e8a31cc54bcf",
   "metadata": {},
   "source": [
    "# Save Gene-Peak Links for relevant genes"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9b6a6553",
   "metadata": {},
   "source": [
    "## Curate list of relevant genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "fa9796e1-e88c-4267-93ed-611d1c378a40",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "total_genes <- c()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "65cbf6df-c508-458d-b771-961d6573c17a",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "### MG0 Marker Genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ba5b8f40-d73c-49f2-a3c7-1b883663d89f",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "\"package 'readxl' was built under R version 4.3.3\"\n",
      "\u001b[1m\u001b[22mNew names:\n",
      "\u001b[36m•\u001b[39m `` -> `...1`\n"
     ]
    }
   ],
   "source": [
    "library(readxl)\n",
    "library(dplyr)\n",
    "\n",
    "# Define file path\n",
    "file_path <- \"../../Annotate_Mic_Subtype/11_clusters/Marker_Genes.xlsx\"\n",
    "\n",
    "# Read the specific sheet \"Cluster_0\"\n",
    "gene_data <- read_excel(file_path, sheet = \"Cluster_0\")\n",
    "\n",
    "# Ensure column names are correctly recognized\n",
    "colnames(gene_data) <- c(\"gene\", \"p_val\", \"avg_log2FC\", \"pct.1\", \"pct.2\", \"p_val_adj\")\n",
    "\n",
    "# Filter genes based on criteria: p_val_adj < 0.001 and avg_log2FC > 0.25\n",
    "filtered_genes <- gene_data %>%\n",
    "  filter(p_val_adj < 0.001, avg_log2FC > 0.25)\n",
    "\n",
    "total_genes <- c(total_genes, as.vector(filtered_genes$gene))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ea2a4f11-a333-48e7-b76b-20ec7101fca4",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "### MG2 Marker Genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7fcb577b-16cf-4b4f-9787-9aa9f53a1f69",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[1m\u001b[22mNew names:\n",
      "\u001b[36m•\u001b[39m `` -> `...1`\n"
     ]
    }
   ],
   "source": [
    "library(readxl)\n",
    "library(dplyr)\n",
    "\n",
    "# Define file path\n",
    "file_path <- \"../../Annotate_Mic_Subtype/11_clusters/Marker_Genes.xlsx\"\n",
    "\n",
    "# Read the specific sheet \"Cluster_0\"\n",
    "gene_data <- read_excel(file_path, sheet = \"Cluster_2\")\n",
    "\n",
    "# Ensure column names are correctly recognized\n",
    "colnames(gene_data) <- c(\"gene\", \"p_val\", \"avg_log2FC\", \"pct.1\", \"pct.2\", \"p_val_adj\")\n",
    "\n",
    "# Filter genes based on criteria: p_val_adj < 0.001 and avg_log2FC > 0.25\n",
    "filtered_genes <- gene_data %>%\n",
    "  filter(p_val_adj < 0.001, avg_log2FC > 0.25)\n",
    "\n",
    "total_genes <- c(total_genes, as.vector(filtered_genes$gene))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1b813fdf-d4b5-42d0-8afa-d0aa0f796c26",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "### MG3 Marker Genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c2d6cd21-e324-4eca-aaa9-74a0d2643ac0",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[1m\u001b[22mNew names:\n",
      "\u001b[36m•\u001b[39m `` -> `...1`\n"
     ]
    }
   ],
   "source": [
    "library(readxl)\n",
    "library(dplyr)\n",
    "\n",
    "# Define file path\n",
    "file_path <- \"../../Annotate_Mic_Subtype/11_clusters/Marker_Genes.xlsx\"\n",
    "\n",
    "# Read the specific sheet \"Cluster_0\"\n",
    "gene_data <- read_excel(file_path, sheet = \"Cluster_3\")\n",
    "\n",
    "# Ensure column names are correctly recognized\n",
    "colnames(gene_data) <- c(\"gene\", \"p_val\", \"avg_log2FC\", \"pct.1\", \"pct.2\", \"p_val_adj\")\n",
    "\n",
    "# Filter genes based on criteria: p_val_adj < 0.001 and avg_log2FC > 0.25\n",
    "filtered_genes <- gene_data %>%\n",
    "  filter(p_val_adj < 0.001, avg_log2FC > 0.25)\n",
    "\n",
    "total_genes <- c(total_genes, as.vector(filtered_genes$gene))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fd901831-51a9-4e27-b0b5-4a46f2179b46",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "### DEGs for MG0,2,3 and DAM"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5d8f371b-d765-4c5d-97bf-1c8fceb0347a",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[1m\u001b[22mNew names:\n",
      "\u001b[36m•\u001b[39m `` -> `...1`\n",
      "\u001b[1m\u001b[22mNew names:\n",
      "\u001b[36m•\u001b[39m `` -> `...1`\n",
      "\u001b[1m\u001b[22mNew names:\n",
      "\u001b[36m•\u001b[39m `` -> `...1`\n"
     ]
    }
   ],
   "source": [
    "# Define file path\n",
    "file_path <- \"../../DEG_GO_Analysis/DEGs_RBD_vs_HC/Significant_DEGs_RBD_vs_HC.xlsx\"\n",
    "\n",
    "# Read the specific sheet \"Cluster_0\"\n",
    "gene_data <- read_excel(file_path, sheet = \"Cluster_0\")\n",
    "# Ensure column names are correctly recognized\n",
    "colnames(gene_data) <- c(\"gene\", \"p_val\", \"avg_log2FC\", \"pct.1\", \"pct.2\", \"p_val_adj\")\n",
    "# Filter genes based on criteria: p_val_adj < 0.001 and avg_log2FC > 0.25\n",
    "filtered_genes <- gene_data %>%\n",
    "  filter(p_val_adj < 0.001, avg_log2FC > 0.25)\n",
    "\n",
    "total_genes <- c(total_genes, as.vector(filtered_genes$gene))\n",
    "\n",
    "\n",
    "gene_data <- read_excel(file_path, sheet = \"Cluster_2\")\n",
    "# Ensure column names are correctly recognized\n",
    "colnames(gene_data) <- c(\"gene\", \"p_val\", \"avg_log2FC\", \"pct.1\", \"pct.2\", \"p_val_adj\")\n",
    "# Filter genes based on criteria: p_val_adj < 0.001 and avg_log2FC > 0.25\n",
    "filtered_genes <- gene_data %>%\n",
    "  filter(p_val_adj < 0.001, avg_log2FC > 0.25)\n",
    "\n",
    "total_genes <- c(total_genes, as.vector(filtered_genes$gene))\n",
    "\n",
    "gene_data <- read_excel(file_path, sheet = \"Cluster_3\")\n",
    "# Ensure column names are correctly recognized\n",
    "colnames(gene_data) <- c(\"gene\", \"p_val\", \"avg_log2FC\", \"pct.1\", \"pct.2\", \"p_val_adj\")\n",
    "# Filter genes based on criteria: p_val_adj < 0.001 and avg_log2FC > 0.25\n",
    "filtered_genes <- gene_data %>%\n",
    "  filter(p_val_adj < 0.001, avg_log2FC > 0.25)\n",
    "\n",
    "total_genes <- c(total_genes, as.vector(filtered_genes$gene))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cb544acc-e159-45b4-b53f-5f84523d9d2f",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[1m\u001b[22mNew names:\n",
      "\u001b[36m•\u001b[39m `` -> `...1`\n",
      "\u001b[1m\u001b[22mNew names:\n",
      "\u001b[36m•\u001b[39m `` -> `...1`\n",
      "\u001b[1m\u001b[22mNew names:\n",
      "\u001b[36m•\u001b[39m `` -> `...1`\n"
     ]
    }
   ],
   "source": [
    "# Define file path\n",
    "file_path <- \"../../DEG_GO_Analysis/DEGs_PD_vs_HC/Significant_DEGs_PD_vs_HC.xlsx\"\n",
    "\n",
    "# Read the specific sheet \"Cluster_0\"\n",
    "gene_data <- read_excel(file_path, sheet = \"Cluster_0\")\n",
    "# Ensure column names are correctly recognized\n",
    "colnames(gene_data) <- c(\"gene\", \"p_val\", \"avg_log2FC\", \"pct.1\", \"pct.2\", \"p_val_adj\")\n",
    "# Filter genes based on criteria: p_val_adj < 0.001 and avg_log2FC > 0.25\n",
    "filtered_genes <- gene_data %>%\n",
    "  filter(p_val_adj < 0.001, avg_log2FC > 0.25)\n",
    "\n",
    "total_genes <- c(total_genes, as.vector(filtered_genes$gene))\n",
    "\n",
    "\n",
    "gene_data <- read_excel(file_path, sheet = \"Cluster_2\")\n",
    "# Ensure column names are correctly recognized\n",
    "colnames(gene_data) <- c(\"gene\", \"p_val\", \"avg_log2FC\", \"pct.1\", \"pct.2\", \"p_val_adj\")\n",
    "# Filter genes based on criteria: p_val_adj < 0.001 and avg_log2FC > 0.25\n",
    "filtered_genes <- gene_data %>%\n",
    "  filter(p_val_adj < 0.001, avg_log2FC > 0.25)\n",
    "\n",
    "total_genes <- c(total_genes, as.vector(filtered_genes$gene))\n",
    "\n",
    "gene_data <- read_excel(file_path, sheet = \"Cluster_3\")\n",
    "# Ensure column names are correctly recognized\n",
    "colnames(gene_data) <- c(\"gene\", \"p_val\", \"avg_log2FC\", \"pct.1\", \"pct.2\", \"p_val_adj\")\n",
    "# Filter genes based on criteria: p_val_adj < 0.001 and avg_log2FC > 0.25\n",
    "filtered_genes <- gene_data %>%\n",
    "  filter(p_val_adj < 0.001, avg_log2FC > 0.25)\n",
    "\n",
    "total_genes <- c(total_genes, as.vector(filtered_genes$gene))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "573a23b8-fc08-4377-baf6-d65d38b33e12",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "dam_ad_genes <- c(\"CD83\", \"NAMPT\", \"HIF1A\", \"SRGN\", \"FTH1\", \"FTL\", \"RPL32\", \"TPT1\", \"RPLP1\", \n",
    "               \"RPS4X\", \"RPS24\", \"RPS11\", \"RPS15\", \"PLEK\", \"CCL3\", \"IL1B\", \"IFIT2\", \"B2M\", \n",
    "               \"CD74\", \"MAMDC2\", \"EEF1B2\", \"PDPN\", \"RPL10A\", \"RPL34\", \"RPL35A\", \"RPL37\", \n",
    "               \"RPL7\", \"RPL38\", \"RPS16\", \"RPL6\", \"RPS14\", \"RPS9\", \"RPL30\", \"APOE\", \"RPS25\", \n",
    "               \"RPS20\", \"EEF1A1\", \"RPS8\", \"RPL13\", \"RPL14\", \"RPL8\", \"RPL35\", \"RPL23\", \"FAU\", \n",
    "               \"RPS19\", \"RPL27A\", \"RPS3\", \"RPL37A\", \"ELOVL5\", \"SOCS6\", \"CADM1\", \n",
    "               \"ITGAX\", \"NAP1L1\", \"CEBPA\", \"TREM2\", \"PLXDC2\", \"SOAT1\", \"LIPA\", \"CD84\", \"CREG1\", \n",
    "               \"FAM20C\", \"SLC16A3\", \"ARHGEF7\", \"GPI\", \"CUX1\", \"DHRS3\", \"CSF2RA\", \"PSAP\", \n",
    "               \"RASSF3\", \"RHEB\", \"BCL2\", \"NFE2L2\", \"TLR2\", \"CHPT1\", \"FGL2\", \"EFR3A\", \"ARAP2\", \n",
    "               \"CLEC7A\", \"CYBA\", \"ITM2B\", \"CD37\", \"ITGB2\", \"ARPC3\", \"HLA-E\", \"EEF2\", \"PABPC1\", \n",
    "               \"NPC2\", \"FCER1G\", \"RPSA\", \"TYROBP\", \"RPL31\", \"EEF1D\", \"RPL12\", \"CHST11\", \"FCGR3A\", \n",
    "               \"AZIN1\", \"MFSD1\", \"GNAS\", \"NOP58\", \"LAT2\", \"CTSZ\", \"SNX3\", \"ADAR\", \"AXL\", \"BST2\", \n",
    "               \"OXCT1\", \"SREBF2\", \"NRP1\", \"USP12\", \"GPNMB\", \"MYO1E\", \"SPP1\", \"ELL2\", \"PADI2\", \n",
    "               \"ATP1B3\", \"RNF149\", \"SLC11A1\", \"KLHL6\", \"PGK1\", \"CTSB\", \"SDCBP\")\n",
    "total_genes <- c(total_genes, as.vector(dam_ad_genes))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "48c66d13-df11-4d10-bfa7-ff5a0bade9f3",
   "metadata": {},
   "source": [
    "### Exc RORB genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cca8a364-aa8f-40ae-80ae-169b6cba2740",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[1m\u001b[22mNew names:\n",
      "\u001b[36m•\u001b[39m `` -> `...1`\n",
      "\u001b[1m\u001b[22mNew names:\n",
      "\u001b[36m•\u001b[39m `` -> `...1`\n",
      "\u001b[1m\u001b[22mNew names:\n",
      "\u001b[36m•\u001b[39m `` -> `...1`\n",
      "\u001b[1m\u001b[22mNew names:\n",
      "\u001b[36m•\u001b[39m `` -> `...1`\n",
      "\u001b[1m\u001b[22mNew names:\n",
      "\u001b[36m•\u001b[39m `` -> `...1`\n",
      "\u001b[1m\u001b[22mNew names:\n",
      "\u001b[36m•\u001b[39m `` -> `...1`\n",
      "\u001b[1m\u001b[22mNew names:\n",
      "\u001b[36m•\u001b[39m `` -> `...1`\n",
      "\u001b[1m\u001b[22mNew names:\n",
      "\u001b[36m•\u001b[39m `` -> `...1`\n",
      "\u001b[1m\u001b[22mNew names:\n",
      "\u001b[36m•\u001b[39m `` -> `...1`\n",
      "\u001b[1m\u001b[22mNew names:\n",
      "\u001b[36m•\u001b[39m `` -> `...1`\n",
      "\u001b[1m\u001b[22mNew names:\n",
      "\u001b[36m•\u001b[39m `` -> `...1`\n",
      "\u001b[1m\u001b[22mNew names:\n",
      "\u001b[36m•\u001b[39m `` -> `...1`\n",
      "\u001b[1m\u001b[22mNew names:\n",
      "\u001b[36m•\u001b[39m `` -> `...1`\n",
      "\u001b[1m\u001b[22mNew names:\n",
      "\u001b[36m•\u001b[39m `` -> `...1`\n",
      "\u001b[1m\u001b[22mNew names:\n",
      "\u001b[36m•\u001b[39m `` -> `...1`\n",
      "\u001b[1m\u001b[22mNew names:\n",
      "\u001b[36m•\u001b[39m `` -> `...1`\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Exc L4-5 RORB LCN15\"\n",
      "[1] \"Exc L3-4 RORB SEMA6D\"\n",
      "[1] \"Exc L3-5 RORB HSPB3\"\n",
      "[1] \"Exc L4-5 RORB HNRNPA1P46\"\n",
      "[1] \"Exc L3-4 RORB FOLH1B\"\n",
      "[1] \"Exc L2-4 RORB GRIK1\"\n",
      "[1] \"Exc L4-5 RORB RPL31P31\"\n",
      "[1] \"Exc L4-5 RORB LINC01474\"\n",
      "[1] \"Exc L3-5 RORB CMAHP\"\n",
      "[1] \"Exc L3-4 RORB PRSS12\"\n",
      "[1] \"Exc L5 RORB LINC01202\"\n",
      "[1] \"Exc L5 RORB SNHG7\"\n",
      "[1] \"Exc L3 RORB CARTPT\"\n",
      "[1] \"Exc L3-5 RORB CD24\"\n",
      "[1] \"Exc L5-6 RORB LINC00320\"\n",
      "[1] \"Exc L4-6 RORB HPCA\"\n"
     ]
    }
   ],
   "source": [
    "library(readxl)\n",
    "library(dplyr)\n",
    "\n",
    "# Define file path\n",
    "file_path <- \"../../Annotate_Exc_Subtype/RORB_cells/Marker_Genes.xlsx\"\n",
    "\n",
    "sheet_names <- excel_sheets(file_path)\n",
    "all_sheets <- lapply(sheet_names, function(sheet) {\n",
    "  read_excel(file_path, sheet = sheet)\n",
    "})\n",
    "names(all_sheets) <- sheet_names\n",
    "\n",
    "for (name in sheet_names){\n",
    "    print(name)\n",
    "    gene_data <- all_sheets[[name]]\n",
    "    # Ensure column names are correctly recognized\n",
    "    colnames(gene_data) <- c(\"gene\", \"p_val\", \"avg_log2FC\", \"pct.1\", \"pct.2\", \"p_val_adj\")\n",
    "\n",
    "    # Filter genes based on criteria: p_val_adj < 0.001 and avg_log2FC > 0.25\n",
    "    filtered_genes <- gene_data %>%\n",
    "      filter(p_val_adj < 0.001, avg_log2FC > 0.25)\n",
    "    \n",
    "    total_genes <- c(total_genes, as.vector(filtered_genes$gene))\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bcf408be-0992-4289-b366-f76646d45e9b",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[1m\u001b[22mNew names:\n",
      "\u001b[36m•\u001b[39m `` -> `...1`\n",
      "\u001b[1m\u001b[22mNew names:\n",
      "\u001b[36m•\u001b[39m `` -> `...1`\n",
      "\u001b[1m\u001b[22mNew names:\n",
      "\u001b[36m•\u001b[39m `` -> `...1`\n",
      "\u001b[1m\u001b[22mNew names:\n",
      "\u001b[36m•\u001b[39m `` -> `...1`\n",
      "\u001b[1m\u001b[22mNew names:\n",
      "\u001b[36m•\u001b[39m `` -> `...1`\n",
      "\u001b[1m\u001b[22mNew names:\n",
      "\u001b[36m•\u001b[39m `` -> `...1`\n",
      "\u001b[1m\u001b[22mNew names:\n",
      "\u001b[36m•\u001b[39m `` -> `...1`\n",
      "\u001b[1m\u001b[22mNew names:\n",
      "\u001b[36m•\u001b[39m `` -> `...1`\n",
      "\u001b[1m\u001b[22mNew names:\n",
      "\u001b[36m•\u001b[39m `` -> `...1`\n",
      "\u001b[1m\u001b[22mNew names:\n",
      "\u001b[36m•\u001b[39m `` -> `...1`\n",
      "\u001b[1m\u001b[22mNew names:\n",
      "\u001b[36m•\u001b[39m `` -> `...1`\n",
      "\u001b[1m\u001b[22mNew names:\n",
      "\u001b[36m•\u001b[39m `` -> `...1`\n",
      "\u001b[1m\u001b[22mNew names:\n",
      "\u001b[36m•\u001b[39m `` -> `...1`\n",
      "\u001b[1m\u001b[22mNew names:\n",
      "\u001b[36m•\u001b[39m `` -> `...1`\n",
      "\u001b[1m\u001b[22mNew names:\n",
      "\u001b[36m•\u001b[39m `` -> `...1`\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Exc L4-5 RORB LCN15\"\n",
      "[1] \"Exc L3-4 RORB SEMA6D\"\n",
      "[1] \"Exc L3-5 RORB HSPB3\"\n",
      "[1] \"Exc L4-5 RORB HNRNPA1P46\"\n",
      "[1] \"Exc L3-4 RORB FOLH1B\"\n",
      "[1] \"Exc L2-4 RORB GRIK1\"\n",
      "[1] \"Exc L4-5 RORB RPL31P31\"\n",
      "[1] \"Exc L4-5 RORB LINC01474\"\n",
      "[1] \"Exc L3-5 RORB CMAHP\"\n",
      "[1] \"Exc L3-4 RORB PRSS12\"\n",
      "[1] \"Exc L5 RORB LINC01202\"\n",
      "[1] \"Exc L5 RORB SNHG7\"\n",
      "[1] \"Exc L3 RORB CARTPT\"\n",
      "[1] \"Exc L3-5 RORB CD24\"\n",
      "[1] \"Exc L5-6 RORB LINC00320\"\n"
     ]
    }
   ],
   "source": [
    "# Define file path\n",
    "file_path <- \"../../DEG_GO_Analysis/DEGs_PD_vs_HC/Significant_DEGs_PD_vs_HC_RORB.xlsx\"\n",
    "\n",
    "sheet_names <- excel_sheets(file_path)\n",
    "all_sheets <- lapply(sheet_names, function(sheet) {\n",
    "  read_excel(file_path, sheet = sheet)\n",
    "})\n",
    "names(all_sheets) <- sheet_names\n",
    "\n",
    "for (name in sheet_names){\n",
    "    print(name)\n",
    "    gene_data <- all_sheets[[name]]\n",
    "    # Ensure column names are correctly recognized\n",
    "    colnames(gene_data) <- c(\"gene\", \"p_val\", \"avg_log2FC\", \"pct.1\", \"pct.2\", \"p_val_adj\")\n",
    "\n",
    "    # Filter genes based on criteria: p_val_adj < 0.001 and avg_log2FC > 0.25\n",
    "    filtered_genes <- gene_data %>%\n",
    "      filter(p_val_adj < 0.001, avg_log2FC > 0.25)\n",
    "    \n",
    "    total_genes <- c(total_genes, as.vector(filtered_genes$gene))\n",
    "\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8696cf05-1598-4b76-9b24-5d2064fe5cdd",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[1m\u001b[22mNew names:\n",
      "\u001b[36m•\u001b[39m `` -> `...1`\n",
      "\u001b[1m\u001b[22mNew names:\n",
      "\u001b[36m•\u001b[39m `` -> `...1`\n",
      "\u001b[1m\u001b[22mNew names:\n",
      "\u001b[36m•\u001b[39m `` -> `...1`\n",
      "\u001b[1m\u001b[22mNew names:\n",
      "\u001b[36m•\u001b[39m `` -> `...1`\n",
      "\u001b[1m\u001b[22mNew names:\n",
      "\u001b[36m•\u001b[39m `` -> `...1`\n",
      "\u001b[1m\u001b[22mNew names:\n",
      "\u001b[36m•\u001b[39m `` -> `...1`\n",
      "\u001b[1m\u001b[22mNew names:\n",
      "\u001b[36m•\u001b[39m `` -> `...1`\n",
      "\u001b[1m\u001b[22mNew names:\n",
      "\u001b[36m•\u001b[39m `` -> `...1`\n",
      "\u001b[1m\u001b[22mNew names:\n",
      "\u001b[36m•\u001b[39m `` -> `...1`\n",
      "\u001b[1m\u001b[22mNew names:\n",
      "\u001b[36m•\u001b[39m `` -> `...1`\n",
      "\u001b[1m\u001b[22mNew names:\n",
      "\u001b[36m•\u001b[39m `` -> `...1`\n",
      "\u001b[1m\u001b[22mNew names:\n",
      "\u001b[36m•\u001b[39m `` -> `...1`\n",
      "\u001b[1m\u001b[22mNew names:\n",
      "\u001b[36m•\u001b[39m `` -> `...1`\n",
      "\u001b[1m\u001b[22mNew names:\n",
      "\u001b[36m•\u001b[39m `` -> `...1`\n",
      "\u001b[1m\u001b[22mNew names:\n",
      "\u001b[36m•\u001b[39m `` -> `...1`\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Exc L4-5 RORB LCN15\"\n",
      "[1] \"Exc L3-4 RORB SEMA6D\"\n",
      "[1] \"Exc L3-5 RORB HSPB3\"\n",
      "[1] \"Exc L4-5 RORB HNRNPA1P46\"\n",
      "[1] \"Exc L3-4 RORB FOLH1B\"\n",
      "[1] \"Exc L2-4 RORB GRIK1\"\n",
      "[1] \"Exc L4-5 RORB RPL31P31\"\n",
      "[1] \"Exc L4-5 RORB LINC01474\"\n",
      "[1] \"Exc L3-5 RORB CMAHP\"\n",
      "[1] \"Exc L3-4 RORB PRSS12\"\n",
      "[1] \"Exc L5 RORB LINC01202\"\n",
      "[1] \"Exc L5 RORB SNHG7\"\n",
      "[1] \"Exc L3 RORB CARTPT\"\n",
      "[1] \"Exc L3-5 RORB CD24\"\n",
      "[1] \"Exc L5-6 RORB LINC00320\"\n"
     ]
    }
   ],
   "source": [
    "# Define file path\n",
    "file_path <- \"../../DEG_GO_Analysis/DEGs_RBD_vs_HC/Significant_DEGs_RBD_vs_HC_RORB.xlsx\"\n",
    "\n",
    "sheet_names <- excel_sheets(file_path)\n",
    "all_sheets <- lapply(sheet_names, function(sheet) {\n",
    "  read_excel(file_path, sheet = sheet)\n",
    "})\n",
    "names(all_sheets) <- sheet_names\n",
    "\n",
    "for (name in sheet_names){\n",
    "    print(name)\n",
    "    gene_data <- all_sheets[[name]]\n",
    "    # Ensure column names are correctly recognized\n",
    "    colnames(gene_data) <- c(\"gene\", \"p_val\", \"avg_log2FC\", \"pct.1\", \"pct.2\", \"p_val_adj\")\n",
    "\n",
    "    # Filter genes based on criteria: p_val_adj < 0.001 and avg_log2FC > 0.25\n",
    "    filtered_genes <- gene_data %>%\n",
    "      filter(p_val_adj < 0.001, avg_log2FC > 0.25)\n",
    "    \n",
    "    total_genes <- c(total_genes, as.vector(filtered_genes$gene))\n",
    "\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3315edd9-4109-4296-ab33-70b07bac5eed",
   "metadata": {},
   "source": [
    "### Inh Subset Genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "98a3209b-ac8f-4acc-9e0d-a2b286c1c596",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[1m\u001b[22mNew names:\n",
      "\u001b[36m•\u001b[39m `` -> `...1`\n",
      "\u001b[1m\u001b[22mNew names:\n",
      "\u001b[36m•\u001b[39m `` -> `...1`\n",
      "\u001b[1m\u001b[22mNew names:\n",
      "\u001b[36m•\u001b[39m `` -> `...1`\n",
      "\u001b[1m\u001b[22mNew names:\n",
      "\u001b[36m•\u001b[39m `` -> `...1`\n",
      "\u001b[1m\u001b[22mNew names:\n",
      "\u001b[36m•\u001b[39m `` -> `...1`\n",
      "\u001b[1m\u001b[22mNew names:\n",
      "\u001b[36m•\u001b[39m `` -> `...1`\n",
      "\u001b[1m\u001b[22mNew names:\n",
      "\u001b[36m•\u001b[39m `` -> `...1`\n",
      "\u001b[1m\u001b[22mNew names:\n",
      "\u001b[36m•\u001b[39m `` -> `...1`\n",
      "\u001b[1m\u001b[22mNew names:\n",
      "\u001b[36m•\u001b[39m `` -> `...1`\n",
      "\u001b[1m\u001b[22mNew names:\n",
      "\u001b[36m•\u001b[39m `` -> `...1`\n",
      "\u001b[1m\u001b[22mNew names:\n",
      "\u001b[36m•\u001b[39m `` -> `...1`\n",
      "\u001b[1m\u001b[22mNew names:\n",
      "\u001b[36m•\u001b[39m `` -> `...1`\n",
      "\u001b[1m\u001b[22mNew names:\n",
      "\u001b[36m•\u001b[39m `` -> `...1`\n",
      "\u001b[1m\u001b[22mNew names:\n",
      "\u001b[36m•\u001b[39m `` -> `...1`\n",
      "\u001b[1m\u001b[22mNew names:\n",
      "\u001b[36m•\u001b[39m `` -> `...1`\n",
      "\u001b[1m\u001b[22mNew names:\n",
      "\u001b[36m•\u001b[39m `` -> `...1`\n",
      "\u001b[1m\u001b[22mNew names:\n",
      "\u001b[36m•\u001b[39m `` -> `...1`\n",
      "\u001b[1m\u001b[22mNew names:\n",
      "\u001b[36m•\u001b[39m `` -> `...1`\n",
      "\u001b[1m\u001b[22mNew names:\n",
      "\u001b[36m•\u001b[39m `` -> `...1`\n",
      "\u001b[1m\u001b[22mNew names:\n",
      "\u001b[36m•\u001b[39m `` -> `...1`\n",
      "\u001b[1m\u001b[22mNew names:\n",
      "\u001b[36m•\u001b[39m `` -> `...1`\n",
      "\u001b[1m\u001b[22mNew names:\n",
      "\u001b[36m•\u001b[39m `` -> `...1`\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Inh L1 SST CXCL14\"\n",
      "[1] \"Inh L1-4 LAMP5 DUSP4\"\n",
      "[1] \"Inh L3-5 SST MAFB\"\n",
      "[1] \"Inh L1 PAX6 CA4\"\n",
      "[1] \"Inh L2-4 PVALB C8orf4\"\n",
      "[1] \"Inh L5-6 PVALB STON2\"\n",
      "[1] \"Inh L1-6 LAMP5 CA13\"\n",
      "[1] \"Inh L6 LAMP5 ANKRD20A11P\"\n",
      "[1] \"Inh L2-4 SST AHR\"\n",
      "[1] \"Inh L5 PVALB CNTNAP3P2\"\n",
      "[1] \"Inh L1 LAMP5 NDNF\"\n",
      "[1] \"Inh L5-6 PVALB FAM150B\"\n",
      "[1] \"Inh L1-3 VIP SSTR1\"\n",
      "[1] \"Inh L1 ADARB2 ADAM33\"\n",
      "[1] \"Inh L4-6 SST MTHFD2P6\"\n",
      "[1] \"Inh L5-6 LAMP5 SFTA3\"\n",
      "[1] \"Inh L1-3 VIP GGH\"\n",
      "[1] \"Inh L1-3 PVALB WFDC2\"\n",
      "[1] \"Inh L5-6 SST ISOC1\"\n",
      "[1] \"Inh L5-6 SST TH\"\n",
      "[1] \"Inh L5-6 SST KLHL14\"\n",
      "[1] \"Inh L1 LAMP5 GGT8P\"\n"
     ]
    }
   ],
   "source": [
    "library(readxl)\n",
    "library(dplyr)\n",
    "\n",
    "# Define file path\n",
    "file_path <- \"../../Annotate_Inh_Subtype/Inh_subset_cells/Marker_Genes.xlsx\"\n",
    "\n",
    "sheet_names <- excel_sheets(file_path)\n",
    "all_sheets <- lapply(sheet_names, function(sheet) {\n",
    "  read_excel(file_path, sheet = sheet)\n",
    "})\n",
    "names(all_sheets) <- sheet_names\n",
    "\n",
    "for (name in sheet_names){\n",
    "    print(name)\n",
    "    gene_data <- all_sheets[[name]]\n",
    "    # Ensure column names are correctly recognized\n",
    "    colnames(gene_data) <- c(\"gene\", \"p_val\", \"avg_log2FC\", \"pct.1\", \"pct.2\", \"p_val_adj\")\n",
    "\n",
    "    # Filter genes based on criteria: p_val_adj < 0.001 and avg_log2FC > 0.25\n",
    "    filtered_genes <- gene_data %>%\n",
    "      filter(p_val_adj < 0.001, avg_log2FC > 2)\n",
    "    \n",
    "    total_genes <- c(total_genes, as.vector(filtered_genes$gene))\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "92bb2cd5-fbf0-403e-bb5a-606d5f977cd5",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[1m\u001b[22mNew names:\n",
      "\u001b[36m•\u001b[39m `` -> `...1`\n",
      "\u001b[1m\u001b[22mNew names:\n",
      "\u001b[36m•\u001b[39m `` -> `...1`\n",
      "\u001b[1m\u001b[22mNew names:\n",
      "\u001b[36m•\u001b[39m `` -> `...1`\n",
      "\u001b[1m\u001b[22mNew names:\n",
      "\u001b[36m•\u001b[39m `` -> `...1`\n",
      "\u001b[1m\u001b[22mNew names:\n",
      "\u001b[36m•\u001b[39m `` -> `...1`\n",
      "\u001b[1m\u001b[22mNew names:\n",
      "\u001b[36m•\u001b[39m `` -> `...1`\n",
      "\u001b[1m\u001b[22mNew names:\n",
      "\u001b[36m•\u001b[39m `` -> `...1`\n",
      "\u001b[1m\u001b[22mNew names:\n",
      "\u001b[36m•\u001b[39m `` -> `...1`\n",
      "\u001b[1m\u001b[22mNew names:\n",
      "\u001b[36m•\u001b[39m `` -> `...1`\n",
      "\u001b[1m\u001b[22mNew names:\n",
      "\u001b[36m•\u001b[39m `` -> `...1`\n",
      "\u001b[1m\u001b[22mNew names:\n",
      "\u001b[36m•\u001b[39m `` -> `...1`\n",
      "\u001b[1m\u001b[22mNew names:\n",
      "\u001b[36m•\u001b[39m `` -> `...1`\n",
      "\u001b[1m\u001b[22mNew names:\n",
      "\u001b[36m•\u001b[39m `` -> `...1`\n",
      "\u001b[1m\u001b[22mNew names:\n",
      "\u001b[36m•\u001b[39m `` -> `...1`\n",
      "\u001b[1m\u001b[22mNew names:\n",
      "\u001b[36m•\u001b[39m `` -> `...1`\n",
      "\u001b[1m\u001b[22mNew names:\n",
      "\u001b[36m•\u001b[39m `` -> `...1`\n",
      "\u001b[1m\u001b[22mNew names:\n",
      "\u001b[36m•\u001b[39m `` -> `...1`\n",
      "\u001b[1m\u001b[22mNew names:\n",
      "\u001b[36m•\u001b[39m `` -> `...1`\n",
      "\u001b[1m\u001b[22mNew names:\n",
      "\u001b[36m•\u001b[39m `` -> `...1`\n",
      "\u001b[1m\u001b[22mNew names:\n",
      "\u001b[36m•\u001b[39m `` -> `...1`\n",
      "\u001b[1m\u001b[22mNew names:\n",
      "\u001b[36m•\u001b[39m `` -> `...1`\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Inh L1 SST CXCL14\"\n",
      "[1] \"Inh L1-4 LAMP5 DUSP4\"\n",
      "[1] \"Inh L3-5 SST MAFB\"\n",
      "[1] \"Inh L1 PAX6 CA4\"\n",
      "[1] \"Inh L2-4 PVALB C8orf4\"\n",
      "[1] \"Inh L5-6 PVALB STON2\"\n",
      "[1] \"Inh L1-6 LAMP5 CA13\"\n",
      "[1] \"Inh L6 LAMP5 ANKRD20A11P\"\n",
      "[1] \"Inh L2-4 SST AHR\"\n",
      "[1] \"Inh L5 PVALB CNTNAP3P2\"\n",
      "[1] \"Inh L1 LAMP5 NDNF\"\n",
      "[1] \"Inh L5-6 PVALB FAM150B\"\n",
      "[1] \"Inh L1-3 VIP SSTR1\"\n",
      "[1] \"Inh L1 ADARB2 ADAM33\"\n",
      "[1] \"Inh L4-6 SST MTHFD2P6\"\n",
      "[1] \"Inh L5-6 LAMP5 SFTA3\"\n",
      "[1] \"Inh L1-3 VIP GGH\"\n",
      "[1] \"Inh L1-3 PVALB WFDC2\"\n",
      "[1] \"Inh L5-6 SST ISOC1\"\n",
      "[1] \"Inh L5-6 SST TH\"\n",
      "[1] \"Inh L5-6 SST KLHL14\"\n"
     ]
    }
   ],
   "source": [
    "# Define file path\n",
    "file_path <- \"../../DEG_GO_Analysis/DEGs_PD_vs_HC/Significant_DEGs_PD_vs_HC_Inh_subset.xlsx\"\n",
    "\n",
    "sheet_names <- excel_sheets(file_path)\n",
    "all_sheets <- lapply(sheet_names, function(sheet) {\n",
    "  read_excel(file_path, sheet = sheet)\n",
    "})\n",
    "names(all_sheets) <- sheet_names\n",
    "\n",
    "for (name in sheet_names){\n",
    "    print(name)\n",
    "    gene_data <- all_sheets[[name]]\n",
    "    # Ensure column names are correctly recognized\n",
    "    colnames(gene_data) <- c(\"gene\", \"p_val\", \"avg_log2FC\", \"pct.1\", \"pct.2\", \"p_val_adj\")\n",
    "\n",
    "    # Filter genes based on criteria: p_val_adj < 0.001 and avg_log2FC > 0.25\n",
    "    filtered_genes <- gene_data %>%\n",
    "      filter(p_val_adj < 0.001, avg_log2FC > 2)\n",
    "    \n",
    "    total_genes <- c(total_genes, as.vector(filtered_genes$gene))\n",
    "\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9fcd332e-6e21-4e51-bff1-74cf6cc415d4",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[1m\u001b[22mNew names:\n",
      "\u001b[36m•\u001b[39m `` -> `...1`\n",
      "\u001b[1m\u001b[22mNew names:\n",
      "\u001b[36m•\u001b[39m `` -> `...1`\n",
      "\u001b[1m\u001b[22mNew names:\n",
      "\u001b[36m•\u001b[39m `` -> `...1`\n",
      "\u001b[1m\u001b[22mNew names:\n",
      "\u001b[36m•\u001b[39m `` -> `...1`\n",
      "\u001b[1m\u001b[22mNew names:\n",
      "\u001b[36m•\u001b[39m `` -> `...1`\n",
      "\u001b[1m\u001b[22mNew names:\n",
      "\u001b[36m•\u001b[39m `` -> `...1`\n",
      "\u001b[1m\u001b[22mNew names:\n",
      "\u001b[36m•\u001b[39m `` -> `...1`\n",
      "\u001b[1m\u001b[22mNew names:\n",
      "\u001b[36m•\u001b[39m `` -> `...1`\n",
      "\u001b[1m\u001b[22mNew names:\n",
      "\u001b[36m•\u001b[39m `` -> `...1`\n",
      "\u001b[1m\u001b[22mNew names:\n",
      "\u001b[36m•\u001b[39m `` -> `...1`\n",
      "\u001b[1m\u001b[22mNew names:\n",
      "\u001b[36m•\u001b[39m `` -> `...1`\n",
      "\u001b[1m\u001b[22mNew names:\n",
      "\u001b[36m•\u001b[39m `` -> `...1`\n",
      "\u001b[1m\u001b[22mNew names:\n",
      "\u001b[36m•\u001b[39m `` -> `...1`\n",
      "\u001b[1m\u001b[22mNew names:\n",
      "\u001b[36m•\u001b[39m `` -> `...1`\n",
      "\u001b[1m\u001b[22mNew names:\n",
      "\u001b[36m•\u001b[39m `` -> `...1`\n",
      "\u001b[1m\u001b[22mNew names:\n",
      "\u001b[36m•\u001b[39m `` -> `...1`\n",
      "\u001b[1m\u001b[22mNew names:\n",
      "\u001b[36m•\u001b[39m `` -> `...1`\n",
      "\u001b[1m\u001b[22mNew names:\n",
      "\u001b[36m•\u001b[39m `` -> `...1`\n",
      "\u001b[1m\u001b[22mNew names:\n",
      "\u001b[36m•\u001b[39m `` -> `...1`\n",
      "\u001b[1m\u001b[22mNew names:\n",
      "\u001b[36m•\u001b[39m `` -> `...1`\n",
      "\u001b[1m\u001b[22mNew names:\n",
      "\u001b[36m•\u001b[39m `` -> `...1`\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Inh L1 SST CXCL14\"\n",
      "[1] \"Inh L1-4 LAMP5 DUSP4\"\n",
      "[1] \"Inh L3-5 SST MAFB\"\n",
      "[1] \"Inh L1 PAX6 CA4\"\n",
      "[1] \"Inh L2-4 PVALB C8orf4\"\n",
      "[1] \"Inh L5-6 PVALB STON2\"\n",
      "[1] \"Inh L1-6 LAMP5 CA13\"\n",
      "[1] \"Inh L6 LAMP5 ANKRD20A11P\"\n",
      "[1] \"Inh L2-4 SST AHR\"\n",
      "[1] \"Inh L5 PVALB CNTNAP3P2\"\n",
      "[1] \"Inh L1 LAMP5 NDNF\"\n",
      "[1] \"Inh L5-6 PVALB FAM150B\"\n",
      "[1] \"Inh L1-3 VIP SSTR1\"\n",
      "[1] \"Inh L1 ADARB2 ADAM33\"\n",
      "[1] \"Inh L4-6 SST MTHFD2P6\"\n",
      "[1] \"Inh L5-6 LAMP5 SFTA3\"\n",
      "[1] \"Inh L1-3 VIP GGH\"\n",
      "[1] \"Inh L1-3 PVALB WFDC2\"\n",
      "[1] \"Inh L5-6 SST ISOC1\"\n",
      "[1] \"Inh L5-6 SST TH\"\n",
      "[1] \"Inh L5-6 SST KLHL14\"\n"
     ]
    }
   ],
   "source": [
    "# Define file path\n",
    "file_path <- \"../../DEG_GO_Analysis/DEGs_RBD_vs_HC/Significant_DEGs_RBD_vs_HC_Inh_subset.xlsx\"\n",
    "\n",
    "sheet_names <- excel_sheets(file_path)\n",
    "all_sheets <- lapply(sheet_names, function(sheet) {\n",
    "  read_excel(file_path, sheet = sheet)\n",
    "})\n",
    "names(all_sheets) <- sheet_names\n",
    "\n",
    "for (name in sheet_names){\n",
    "    print(name)\n",
    "    gene_data <- all_sheets[[name]]\n",
    "    # Ensure column names are correctly recognized\n",
    "    colnames(gene_data) <- c(\"gene\", \"p_val\", \"avg_log2FC\", \"pct.1\", \"pct.2\", \"p_val_adj\")\n",
    "\n",
    "    # Filter genes based on criteria: p_val_adj < 0.001 and avg_log2FC > 0.25\n",
    "    filtered_genes <- gene_data %>%\n",
    "      filter(p_val_adj < 0.001, avg_log2FC > 2)\n",
    "    \n",
    "    total_genes <- c(total_genes, as.vector(filtered_genes$gene))\n",
    "\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "502219d8-501b-4663-8b6b-e9f4e8e083b9",
   "metadata": {},
   "source": [
    "## Save expression for curated list of genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "956d3961-568a-4028-9288-a2c582829ddf",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "total_genes <- unique(total_genes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "7b5a9180-e1d1-4f1c-af0d-b9718b47f4d0",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "3081"
      ],
      "text/latex": [
       "3081"
      ],
      "text/markdown": [
       "3081"
      ],
      "text/plain": [
       "[1] 3081"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "length(unique(total_genes))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8bd62572-105a-4363-a7f0-0647d4fd24f8",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "# writeLines(total_genes, \"../../Signac_plots/saved_files/mic_total_genes.txt\")\n",
    "# writeLines(total_genes, \"../../Signac_plots/saved_files/exn_rorb_total_genes.txt\")\n",
    "writeLines(total_genes, \"../../Signac_plots/saved_files/inh_subset_total_genes.txt\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "02d262d9-ed1e-478c-aeee-a6d09f199b48",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "# total_genes <- readLines(\"../../Signac_plots/saved_files/mic_total_genes.txt\")\n",
    "# total_genes <- readLines(\"../../Signac_plots/saved_files/exn_rorb_total_genes.txt\")\n",
    "total_genes <- readLines(\"../../Signac_plots/saved_files/inh_subset_total_genes.txt\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6acc304e-9212-45dd-ba10-697564dc5305",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Testing 1679 genes and 138047 peaks\n",
      "\n",
      "Found gene coordinates for 1212 genes\n",
      "\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"Each of the 2 combined objects has sequence levels not in the other:\n",
      "  - in 'x': GL000194.1, GL000195.1, GL000205.2, GL000218.1, GL000219.1, KI270711.1, KI270713.1, KI270721.1, KI270726.1, KI270727.1, KI270728.1, KI270731.1, KI270734.1, GL000009.2, GL000213.1\n",
      "  - in 'y': chrMT\n",
      "  Make sure to always combine/compare objects based on the same reference\n",
      "  genome (use suppressWarnings() to suppress this warning).\"\n"
     ]
    }
   ],
   "source": [
    "# 1.2 Pull out the links that LinkPeaks() created\n",
    "## each element of the GRanges object is one peak; the linked gene lives in metadata\n",
    "ifnb <- LinkPeaks(\n",
    "  object = ifnb,\n",
    "  peak.assay = \"ATAC\",\n",
    "  expression.assay = \"SCT\",\n",
    "  genes.use = total_genes\n",
    ")\n",
    "\n",
    "peak_gene_links <- Links(ifnb[[\"ATAC\"]])\n",
    "df <- as.data.frame(peak_gene_links)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "44419373-7e9f-4be1-8cad-dee7350f6e22",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 6 × 10</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>seqnames</th><th scope=col>start</th><th scope=col>end</th><th scope=col>width</th><th scope=col>strand</th><th scope=col>score</th><th scope=col>gene</th><th scope=col>peak</th><th scope=col>zscore</th><th scope=col>pvalue</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1</th><td>chr1</td><td>  812991</td><td> 1309609</td><td>496619</td><td>*</td><td>0.05285895</td><td>ACAP3 </td><td>chr1-812528-813454    </td><td>4.369249</td><td>6.233726e-06</td></tr>\n",
       "\t<tr><th scope=row>2</th><td>chr1</td><td> 1000172</td><td> 1250407</td><td>250236</td><td>*</td><td>0.05096074</td><td>HES4  </td><td>chr1-1249805-1251009  </td><td>4.138091</td><td>1.751036e-05</td></tr>\n",
       "\t<tr><th scope=row>3</th><td>chr1</td><td> 9088478</td><td> 9250223</td><td>161746</td><td>*</td><td>0.05143187</td><td>SLC2A5</td><td>chr1-9249751-9250695  </td><td>4.781409</td><td>8.703533e-07</td></tr>\n",
       "\t<tr><th scope=row>4</th><td>chr1</td><td> 9088478</td><td> 9438513</td><td>350036</td><td>*</td><td>0.07813180</td><td>SLC2A5</td><td>chr1-9437929-9439097  </td><td>7.460428</td><td>4.312092e-14</td></tr>\n",
       "\t<tr><th scope=row>5</th><td>chr1</td><td> 9651732</td><td> 9689242</td><td> 37511</td><td>*</td><td>0.05129150</td><td>PIK3CD</td><td>chr1-9688541-9689942  </td><td>4.337663</td><td>7.200301e-06</td></tr>\n",
       "\t<tr><th scope=row>6</th><td>chr1</td><td>13583913</td><td>13892792</td><td>308880</td><td>*</td><td>0.05089589</td><td>KAZN  </td><td>chr1-13583110-13584716</td><td>2.885514</td><td>1.953872e-03</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 6 × 10\n",
       "\\begin{tabular}{r|llllllllll}\n",
       "  & seqnames & start & end & width & strand & score & gene & peak & zscore & pvalue\\\\\n",
       "  & <fct> & <int> & <int> & <int> & <fct> & <dbl> & <chr> & <chr> & <dbl> & <dbl>\\\\\n",
       "\\hline\n",
       "\t1 & chr1 &   812991 &  1309609 & 496619 & * & 0.05285895 & ACAP3  & chr1-812528-813454     & 4.369249 & 6.233726e-06\\\\\n",
       "\t2 & chr1 &  1000172 &  1250407 & 250236 & * & 0.05096074 & HES4   & chr1-1249805-1251009   & 4.138091 & 1.751036e-05\\\\\n",
       "\t3 & chr1 &  9088478 &  9250223 & 161746 & * & 0.05143187 & SLC2A5 & chr1-9249751-9250695   & 4.781409 & 8.703533e-07\\\\\n",
       "\t4 & chr1 &  9088478 &  9438513 & 350036 & * & 0.07813180 & SLC2A5 & chr1-9437929-9439097   & 7.460428 & 4.312092e-14\\\\\n",
       "\t5 & chr1 &  9651732 &  9689242 &  37511 & * & 0.05129150 & PIK3CD & chr1-9688541-9689942   & 4.337663 & 7.200301e-06\\\\\n",
       "\t6 & chr1 & 13583913 & 13892792 & 308880 & * & 0.05089589 & KAZN   & chr1-13583110-13584716 & 2.885514 & 1.953872e-03\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 6 × 10\n",
       "\n",
       "| <!--/--> | seqnames &lt;fct&gt; | start &lt;int&gt; | end &lt;int&gt; | width &lt;int&gt; | strand &lt;fct&gt; | score &lt;dbl&gt; | gene &lt;chr&gt; | peak &lt;chr&gt; | zscore &lt;dbl&gt; | pvalue &lt;dbl&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| 1 | chr1 |   812991 |  1309609 | 496619 | * | 0.05285895 | ACAP3  | chr1-812528-813454     | 4.369249 | 6.233726e-06 |\n",
       "| 2 | chr1 |  1000172 |  1250407 | 250236 | * | 0.05096074 | HES4   | chr1-1249805-1251009   | 4.138091 | 1.751036e-05 |\n",
       "| 3 | chr1 |  9088478 |  9250223 | 161746 | * | 0.05143187 | SLC2A5 | chr1-9249751-9250695   | 4.781409 | 8.703533e-07 |\n",
       "| 4 | chr1 |  9088478 |  9438513 | 350036 | * | 0.07813180 | SLC2A5 | chr1-9437929-9439097   | 7.460428 | 4.312092e-14 |\n",
       "| 5 | chr1 |  9651732 |  9689242 |  37511 | * | 0.05129150 | PIK3CD | chr1-9688541-9689942   | 4.337663 | 7.200301e-06 |\n",
       "| 6 | chr1 | 13583913 | 13892792 | 308880 | * | 0.05089589 | KAZN   | chr1-13583110-13584716 | 2.885514 | 1.953872e-03 |\n",
       "\n"
      ],
      "text/plain": [
       "  seqnames start    end      width  strand score      gene  \n",
       "1 chr1       812991  1309609 496619 *      0.05285895 ACAP3 \n",
       "2 chr1      1000172  1250407 250236 *      0.05096074 HES4  \n",
       "3 chr1      9088478  9250223 161746 *      0.05143187 SLC2A5\n",
       "4 chr1      9088478  9438513 350036 *      0.07813180 SLC2A5\n",
       "5 chr1      9651732  9689242  37511 *      0.05129150 PIK3CD\n",
       "6 chr1     13583913 13892792 308880 *      0.05089589 KAZN  \n",
       "  peak                   zscore   pvalue      \n",
       "1 chr1-812528-813454     4.369249 6.233726e-06\n",
       "2 chr1-1249805-1251009   4.138091 1.751036e-05\n",
       "3 chr1-9249751-9250695   4.781409 8.703533e-07\n",
       "4 chr1-9437929-9439097   7.460428 4.312092e-14\n",
       "5 chr1-9688541-9689942   4.337663 7.200301e-06\n",
       "6 chr1-13583110-13584716 2.885514 1.953872e-03"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2fc5e7a2-ce7b-4ea9-a5c3-52689cbd0c05",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "links_df <- as_tibble(df) |>\n",
    "  transmute(\n",
    "    chrom      = seqnames,          # chromosome\n",
    "    chromStart = start              # 0-based for BED\n",
    "      - 1L,                         # Signac stores 1-based; BED wants 0-based half-open\n",
    "    chromEnd   = end,\n",
    "    gene       = gene,\n",
    "    score      = score              # optional\n",
    "  )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f2ae7f83-28c2-46f6-b35d-0cc80cdcc7fd",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "links_df <- links_df |>\n",
    "  mutate(peak_id = paste0(chrom, \":\", chromStart, \"-\", chromEnd))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "edbb6155-a66d-48f4-ad14-80c0410cc1e0",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "library(tidyverse)\n",
    "# write_tsv(links_df,  \"../../Signac_plots/saved_files/HC_gene_peak_links_mic.tsv\")\n",
    "# write_tsv(links_df,  \"../../Signac_plots/saved_files/HC_gene_peak_links_exn_rorb.tsv\")\n",
    "write_tsv(links_df,  \"../../Signac_plots/saved_files/HC_gene_peak_links_inh_subset.tsv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "10c66cca-d984-485f-b7bd-9eedda67f5bd",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "peaks_df <- links_df %>% \n",
    "  select(chrom, chromStart, chromEnd, peak_id)\n",
    "\n",
    "# readr::write_tsv(\n",
    "#   peaks_df, \n",
    "#   \"../../Signac_plots/saved_files/HC_peaks_mic.bed\", \n",
    "#   col_names = FALSE\n",
    "# )\n",
    "\n",
    "# readr::write_tsv(\n",
    "#   peaks_df, \n",
    "#   \"../../Signac_plots/saved_files/HC_peaks_exn_rorb.bed\", \n",
    "#   col_names = FALSE\n",
    "# )\n",
    "\n",
    "readr::write_tsv(\n",
    "  peaks_df, \n",
    "  \"../../Signac_plots/saved_files/HC_peaks_inh_subset.bed\", \n",
    "  col_names = FALSE\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bcf545c6-ec1c-4d1b-afb3-c026040df542",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "tfs <- read.table(\"../../Signac_plots/saved_files/all_tfs.txt\", stringsAsFactors = FALSE)[,1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "70d1eb85-0fc2-468b-8938-a0248e6f8a06",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "DefaultAssay(ifnb) <- \"SCT\"\n",
    "ifnb"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "58da5353-02fd-4ffe-8041-d9fe9acfbd4a",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "# after defining TFs in Python, read them back or compute in R the same way\n",
    "DefaultAssay(ifnb) <- \"SCT\"\n",
    "keep_genes <- union(VariableFeatures(ifnb), tfs)\n",
    "ifnb_subset <- subset(ifnb, features = keep_genes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7eb31ee9-a96b-420a-a3c2-0c23ca20ecb3",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "length(unique(keep_genes))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "45b0f0ea-36af-40ed-8acb-54377ee47f14",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "ifnb_subset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a04f0c9c-bf91-4318-99d8-2240cfd1d73d",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "# SaveLoom(\n",
    "#   object   = ifnb_subset,\n",
    "#   filename = \"../../Signac_plots/saved_files/hc_mic_all_tfs_rna_MIC.loom\",\n",
    "#   assay    = \"SCT\",\n",
    "#   slot     = \"data\"\n",
    "# )\n",
    "\n",
    "# SaveLoom(\n",
    "#   object   = ifnb_subset,\n",
    "#   filename = \"../../Signac_plots/saved_files/hc_exn_all_tfs_rna_EXN_RORB.loom\",\n",
    "#   assay    = \"SCT\",\n",
    "#   slot     = \"data\"\n",
    "# )\n",
    "\n",
    "SaveLoom(\n",
    "  object   = ifnb_subset,\n",
    "  filename = \"../../Signac_plots/saved_files/hc_inh_all_tfs_rna_INH_SUBSET.loom\",\n",
    "  assay    = \"SCT\",\n",
    "  slot     = \"data\"\n",
    ")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.3.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
