{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "6854f228-5efa-450d-9b13-33bf0cc087d2",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "“package ‘Seurat’ was built under R version 4.3.3”\n",
      "Loading required package: SeuratObject\n",
      "\n",
      "Warning message:\n",
      "“package ‘SeuratObject’ was built under R version 4.3.3”\n",
      "Loading required package: sp\n",
      "\n",
      "Warning message:\n",
      "“package ‘sp’ was built under R version 4.3.3”\n",
      "\n",
      "Attaching package: ‘SeuratObject’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    intersect, t\n",
      "\n",
      "\n",
      "Warning message:\n",
      "“package ‘ggplot2’ was built under R version 4.3.3”\n",
      "Welcome to enrichR\n",
      "Checking connection ... \n",
      "\n",
      "Enrichr ... \n",
      "Connection is Live!\n",
      "\n",
      "FlyEnrichr ... \n",
      "Connection is Live!\n",
      "\n",
      "WormEnrichr ... \n",
      "Connection is Live!\n",
      "\n",
      "YeastEnrichr ... \n",
      "Connection is Live!\n",
      "\n",
      "FishEnrichr ... \n",
      "Connection is Live!\n",
      "\n",
      "OxEnrichr ... \n",
      "Connection is Live!\n",
      "\n",
      "Warning message:\n",
      "“package ‘stringr’ was built under R version 4.3.3”\n",
      "Warning message:\n",
      "“package ‘RColorBrewer’ was built under R version 4.3.3”\n",
      "Warning message:\n",
      "“package ‘dplyr’ was built under R version 4.3.3”\n",
      "\n",
      "Attaching package: ‘dplyr’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    filter, lag\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    intersect, setdiff, setequal, union\n",
      "\n",
      "\n",
      "Warning message:\n",
      "“package ‘scales’ was built under R version 4.3.3”\n",
      "Loading required package: grid\n",
      "\n",
      "========================================\n",
      "ComplexHeatmap version 2.18.0\n",
      "Bioconductor page: http://bioconductor.org/packages/ComplexHeatmap/\n",
      "Github page: https://github.com/jokergoo/ComplexHeatmap\n",
      "Documentation: http://jokergoo.github.io/ComplexHeatmap-reference\n",
      "\n",
      "If you use it in published research, please cite either one:\n",
      "- Gu, Z. Complex Heatmap Visualization. iMeta 2022.\n",
      "- Gu, Z. Complex heatmaps reveal patterns and correlations in multidimensional \n",
      "    genomic data. Bioinformatics 2016.\n",
      "\n",
      "\n",
      "The new InteractiveComplexHeatmap package can directly export static \n",
      "complex heatmaps into an interactive Shiny app with zero effort. Have a try!\n",
      "\n",
      "This message can be suppressed by:\n",
      "  suppressPackageStartupMessages(library(ComplexHeatmap))\n",
      "========================================\n",
      "! pheatmap() has been masked by ComplexHeatmap::pheatmap(). Most of the arguments\n",
      "   in the original pheatmap() are identically supported in the new function. You \n",
      "   can still use the original function by explicitly calling pheatmap::pheatmap().\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘ComplexHeatmap’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:pheatmap’:\n",
      "\n",
      "    pheatmap\n",
      "\n",
      "\n",
      "========================================\n",
      "circlize version 0.4.16\n",
      "CRAN page: https://cran.r-project.org/package=circlize\n",
      "Github page: https://github.com/jokergoo/circlize\n",
      "Documentation: https://jokergoo.github.io/circlize_book/book/\n",
      "\n",
      "If you use it in published research, please cite:\n",
      "Gu, Z. circlize implements and enhances circular visualization\n",
      "  in R. Bioinformatics 2014.\n",
      "\n",
      "This message can be suppressed by:\n",
      "  suppressPackageStartupMessages(library(circlize))\n",
      "========================================\n",
      "\n",
      "\n",
      "Warning message:\n",
      "“package ‘tidyr’ was built under R version 4.3.3”\n",
      "Warning message:\n",
      "“package ‘tibble’ was built under R version 4.3.3”\n"
     ]
    }
   ],
   "source": [
    "library(Seurat)\n",
    "library(ggplot2)\n",
    "library(enrichR)\n",
    "library(stringr)\n",
    "library(RColorBrewer)\n",
    "library(pheatmap)\n",
    "library(dplyr)\n",
    "library(scales)\n",
    "library(dplyr)\n",
    "library(ComplexHeatmap)\n",
    "library(circlize)\n",
    "library(tidyr)\n",
    "library(tibble)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a53b4ca9-a86b-4300-863c-fa54b5e6ef45",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "ifnb = readRDS(\"hc_rbd_pd_joint_integrated.rds\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "ee5ae48b-15a5-4f8d-884a-6b8b980680d0",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "ifnb_testct <- ifnb"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "bc62cb72-abe1-425c-b096-5e01ab72e0a5",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " chr [1:231216] \"excitatory neurons\" \"oligodendrocyte precursor cells\" ...\n"
     ]
    }
   ],
   "source": [
    "glimpse(ifnb_testct@meta.data$celltype)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "00e83deb-7ee5-46c8-8b23-4701edf01fcf",
   "metadata": {},
   "source": [
    "# DEG analysis (PD vs HC) for each cell type"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "29298339-558c-4eb5-8f5a-1ae2aaf0c127",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "set.seed(384)\n",
    "Idents(ifnb_testct) <- \"celltype\"\n",
    "deg_results_by_cluster <- list()\n",
    "\n",
    "\n",
    "for (ct in unique(ifnb_testct$celltype)) {\n",
    "  ct_subset <- subset(ifnb_testct, idents = ct)\n",
    "  Idents(ct_subset) <- \"disease\"\n",
    "  \n",
    "  # Check if both \"PD\" and \"HC\" are present in the subset\n",
    "  if (all(c(\"PD\", \"HC\") %in% unique(Idents(ct_subset)))) {\n",
    "    # Perform differential expression analysis between \"PD\" and \"HC\" within this cluster\n",
    "    deg_results <- FindMarkers(ct_subset, ident.1 = \"PD\", ident.2 = \"HC\")\n",
    "    deg_results_by_cluster[[ct]] <- deg_results\n",
    "  } else {\n",
    "    message(paste(\"Skipping ct\", ct, \"- does not contain both PD and HC cells\"))\n",
    "  }\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5e2160f4-49ad-4970-9fa4-969a13d4026c",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "library(openxlsx)\n",
    "\n",
    "# Create a new workbook\n",
    "wb <- createWorkbook()\n",
    "\n",
    "for (ct in unique(ifnb_testct$celltype)) {\n",
    "    if (ct != \"pericytes\"){\n",
    "        # Filter the data for each cluster\n",
    "        cluster_data <- deg_results_by_cluster[[ct]] %>% filter(p_val_adj < 0.05) %>% arrange(desc(avg_log2FC))\n",
    "        \n",
    "        # Add a worksheet with the cluster name\n",
    "        addWorksheet(wb, sheetName = ct)\n",
    "        \n",
    "        # Write the filtered data to the worksheet\n",
    "        writeData(wb, sheet = ct, cluster_data, rowNames = TRUE)\n",
    "    }\n",
    "}\n",
    "\n",
    "# Save the workbook to an Excel file\n",
    "saveWorkbook(wb, file = \"../../DEG_GO_Analysis/DEGs_PD_vs_HC/Significant_DEGs_PD_vs_HC_ct.xlsx\", overwrite = TRUE)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ebab9b71-10b0-4414-81f8-cfc837db09f3",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'excitatory neurons'</li><li>'oligodendrocyte precursor cells'</li><li>'oligodendrocytes'</li><li>'inhibitory neurons'</li><li>'astrocytes'</li><li>'microglia'</li><li>'endothelia cells'</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 'excitatory neurons'\n",
       "\\item 'oligodendrocyte precursor cells'\n",
       "\\item 'oligodendrocytes'\n",
       "\\item 'inhibitory neurons'\n",
       "\\item 'astrocytes'\n",
       "\\item 'microglia'\n",
       "\\item 'endothelia cells'\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 'excitatory neurons'\n",
       "2. 'oligodendrocyte precursor cells'\n",
       "3. 'oligodendrocytes'\n",
       "4. 'inhibitory neurons'\n",
       "5. 'astrocytes'\n",
       "6. 'microglia'\n",
       "7. 'endothelia cells'\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[1] \"excitatory neurons\"              \"oligodendrocyte precursor cells\"\n",
       "[3] \"oligodendrocytes\"                \"inhibitory neurons\"             \n",
       "[5] \"astrocytes\"                      \"microglia\"                      \n",
       "[7] \"endothelia cells\"               "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "library(openxlsx)\n",
    "\n",
    "xlsx_path <- \"../../DEG_GO_Analysis/DEGs_PD_vs_HC/Significant_DEGs_PD_vs_HC_ct.xlsx\"\n",
    "\n",
    "# Get sheet names (these become the sublist names)\n",
    "sheets <- getSheetNames(xlsx_path)\n",
    "\n",
    "# Optionally drop any sheet you don't want\n",
    "sheets <- setdiff(sheets, \"pericytes\")\n",
    "\n",
    "# Read each sheet -> data.frame with gene rownames\n",
    "deg_results_by_cluster <- setNames(\n",
    "  lapply(sheets, function(sh) {\n",
    "    df <- read.xlsx(xlsx_path, sheet = sh, rowNames = TRUE)\n",
    "\n",
    "    # Make sure numeric columns are numeric (robust to minor name variants)\n",
    "    nm <- tolower(names(df))\n",
    "    num_cols <- names(df)[nm %in% c(\"p_val\",\"avg_log2fc\",\"avg_log2_fc\",\"avg_logfc\",\n",
    "                                    \"pct.1\",\"pct1\",\"pct.2\",\"pct2\",\"p_val_adj\",\"padj\",\"p_adj\")]\n",
    "    for (cc in num_cols) df[[cc]] <- as.numeric(df[[cc]])\n",
    "\n",
    "    df\n",
    "  }),\n",
    "  sheets\n",
    ")\n",
    "\n",
    "# quick checks\n",
    "names(deg_results_by_cluster)\n",
    "# e.g. deg_results_by_cluster[[\"astrocytes\"]][1:5, ]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "94f0d53b",
   "metadata": {},
   "source": [
    "### Dot plot of top DEGs per cell type"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "270d426f-947f-4f0f-9989-92380f8782fb",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Top 10/20 genes for cell type: excitatory neurons \n",
      "                 p_val avg_log2FC pct.1 pct.2     p_val_adj     gene\n",
      "CRNDE     4.537810e-15   2.321718 0.014 0.006  1.660884e-10    CRNDE\n",
      "HIST1H1D  3.169666e-24   2.240203 0.010 0.002  1.160129e-19 HIST1H1D\n",
      "MET      8.520966e-124   1.739333 0.098 0.036 3.118759e-119      MET\n",
      "PRKY     4.600518e-246   1.713994 0.152 0.048 1.683836e-241     PRKY\n",
      "NLGN4Y    0.000000e+00   1.569057 0.615 0.207  0.000000e+00   NLGN4Y\n",
      "HLA-DMA   1.118747e-57   1.551068 0.077 0.037  4.094724e-53  HLA-DMA\n",
      "PTGDR     2.083627e-15   1.537189 0.025 0.013  7.626283e-11    PTGDR\n",
      "RPS4Y1   4.502661e-241   1.507591 0.180 0.065 1.648019e-236   RPS4Y1\n",
      "ZFY      1.781120e-273   1.478723 0.204 0.075 6.519076e-269      ZFY\n",
      "ZFY-AS1   2.801526e-20   1.453695 0.017 0.007  1.025387e-15  ZFY-AS1\n",
      "CARTPT    1.363782e-20   1.427322 0.028 0.014  4.991580e-16   CARTPT\n",
      "EIF1AY    2.426816e-56   1.404657 0.045 0.016  8.882388e-52   EIF1AY\n",
      "KDM5D     0.000000e+00   1.371192 0.340 0.125  0.000000e+00    KDM5D\n",
      "COX7A1    3.975901e-40   1.367478 0.044 0.019  1.455220e-35   COX7A1\n",
      "HIST1H4C 1.344204e-116   1.356724 0.115 0.048 4.919921e-112 HIST1H4C\n",
      "USP9Y     0.000000e+00   1.336385 0.614 0.210  0.000000e+00    USP9Y\n",
      "ANO2      2.688327e-30   1.243594 0.127 0.089  9.839547e-26     ANO2\n",
      "UTY       0.000000e+00   1.235027 0.589 0.206  0.000000e+00      UTY\n",
      "PCDH11Y  1.348527e-223   1.224744 0.208 0.088 4.935745e-219  PCDH11Y\n",
      "ASB17     1.432956e-22   1.222335 0.019 0.008  5.244762e-18    ASB17\n",
      "\n",
      "Top 10/20 genes for cell type: oligodendrocyte precursor cells \n",
      "                p_val avg_log2FC pct.1 pct.2     p_val_adj    gene\n",
      "CCKAR    1.254318e-09   2.717930 0.017 0.003  4.590930e-05   CCKAR\n",
      "ISG15    7.824666e-17   2.108950 0.059 0.021  2.863906e-12   ISG15\n",
      "IFIT1    8.739538e-27   1.783889 0.111 0.044  3.198758e-22   IFIT1\n",
      "PCDH11Y 3.437982e-115   1.609913 0.395 0.154 1.258336e-110 PCDH11Y\n",
      "NLGN4Y  2.533474e-219   1.425755 0.666 0.275 9.272767e-215  NLGN4Y\n",
      "DDX3Y    9.146755e-39   1.323051 0.125 0.040  3.347804e-34   DDX3Y\n",
      "OAS1     1.222436e-06   1.298570 0.029 0.013  4.474238e-02    OAS1\n",
      "EPSTI1   1.460167e-07   1.217921 0.048 0.024  5.344356e-03  EPSTI1\n",
      "KDM5D    2.945398e-51   1.214511 0.210 0.081  1.078045e-46   KDM5D\n",
      "ZFY      1.975166e-41   1.197330 0.164 0.062  7.229306e-37     ZFY\n",
      "\n",
      "Top 10/20 genes for cell type: oligodendrocytes \n",
      "                p_val avg_log2FC pct.1 pct.2     p_val_adj    gene\n",
      "DDX3Y    0.000000e+00   2.089254 0.190 0.039  0.000000e+00   DDX3Y\n",
      "FERMT3   8.871433e-88   1.587876 0.031 0.010  3.247033e-83  FERMT3\n",
      "NPY2R   1.158936e-156   1.564565 0.026 0.003 4.241823e-152   NPY2R\n",
      "ZFY      0.000000e+00   1.506705 0.092 0.025  0.000000e+00     ZFY\n",
      "TMSB4Y   8.630577e-39   1.394003 0.011 0.003  3.158877e-34  TMSB4Y\n",
      "RPS4Y1  4.312671e-141   1.351810 0.047 0.014 1.578481e-136  RPS4Y1\n",
      "TLR4     4.982291e-85   1.269998 0.035 0.013  1.823568e-80    TLR4\n",
      "COL27A1  9.400289e-17   1.258259 0.012 0.006  3.440600e-12 COL27A1\n",
      "MTHFD2   5.655929e-55   1.245262 0.020 0.007  2.070127e-50  MTHFD2\n",
      "EIF1AY  1.399038e-178   1.239103 0.062 0.019 5.120620e-174  EIF1AY\n",
      "\n",
      "Top 10/20 genes for cell type: inhibitory neurons \n",
      "                    p_val avg_log2FC pct.1 pct.2     p_val_adj        gene\n",
      "PCDH11Y     1.760836e-236  1.5036201 0.311 0.137 6.444836e-232     PCDH11Y\n",
      "USP9Y        0.000000e+00  1.1505521 0.566 0.241  0.000000e+00       USP9Y\n",
      "NLGN4Y       0.000000e+00  1.1412778 0.558 0.237  0.000000e+00      NLGN4Y\n",
      "KDM5D       4.781876e-166  1.1212556 0.260 0.118 1.750214e-161       KDM5D\n",
      "UTY          0.000000e+00  0.8831736 0.521 0.236  0.000000e+00         UTY\n",
      "FLG-AS1      8.807220e-65  0.5472773 0.347 0.241  3.223531e-60     FLG-AS1\n",
      "FAM153B      2.429814e-49  0.5029815 0.334 0.240  8.893361e-45     FAM153B\n",
      "DLX6-AS1     7.027532e-86  0.4917045 0.688 0.579  2.572147e-81    DLX6-AS1\n",
      "HSP90AA1     3.307716e-39  0.4428614 0.642 0.579  1.210657e-34    HSP90AA1\n",
      "CLU          3.462940e-24  0.4407182 0.410 0.342  1.267471e-19         CLU\n",
      "DNAJA1       2.958877e-41  0.4399726 0.279 0.197  1.082978e-36      DNAJA1\n",
      "PCDH19       8.931320e-37  0.4143120 0.376 0.294  3.268953e-32      PCDH19\n",
      "OLFM3        6.058081e-35  0.4098103 0.495 0.418  2.217318e-30       OLFM3\n",
      "NXPH2        6.499035e-28  0.4094171 0.343 0.277  2.378712e-23       NXPH2\n",
      "TENM2       1.719761e-105  0.4062556 0.907 0.796 6.294495e-101       TENM2\n",
      "ANKRD36     2.824635e-136  0.4047748 0.888 0.810 1.033845e-131     ANKRD36\n",
      "PDE3A        1.237889e-26  0.4017727 0.360 0.294  4.530798e-22       PDE3A\n",
      "C3orf67-AS1  7.467851e-28  0.3967397 0.283 0.217  2.733308e-23 C3orf67-AS1\n",
      "SEZ6L2       1.807873e-31  0.3753669 0.299 0.225  6.616997e-27      SEZ6L2\n",
      "GDA          6.407204e-34  0.3750080 0.267 0.194  2.345101e-29         GDA\n",
      "\n",
      "Top 10/20 genes for cell type: astrocytes \n",
      "                   p_val avg_log2FC pct.1 pct.2     p_val_adj       gene\n",
      "KCNIP4      8.849946e-54  1.4926918 0.365 0.279  3.239169e-49     KCNIP4\n",
      "CSMD1       2.362455e-78  1.3653223 0.301 0.200  8.646821e-74      CSMD1\n",
      "NLGN4Y      0.000000e+00  1.3311865 0.402 0.144  0.000000e+00     NLGN4Y\n",
      "USP9Y       0.000000e+00  1.1712783 0.436 0.169  0.000000e+00      USP9Y\n",
      "UTY         0.000000e+00  1.1662438 0.571 0.233  0.000000e+00        UTY\n",
      "STXBP5L     7.771548e-92  1.0807382 0.339 0.229  2.844464e-87    STXBP5L\n",
      "RBFOX1      3.005514e-77  1.0550028 0.356 0.251  1.100048e-72     RBFOX1\n",
      "SYT1        4.913735e-49  0.9554333 0.273 0.196  1.798476e-44       SYT1\n",
      "ST6GALNAC5 1.133271e-120  0.9350938 0.297 0.174 4.147886e-116 ST6GALNAC5\n",
      "CNTNAP2     4.861921e-60  0.9317256 0.323 0.232  1.779512e-55    CNTNAP2\n",
      "\n",
      "Top 10/20 genes for cell type: microglia \n",
      "               p_val avg_log2FC pct.1 pct.2    p_val_adj    gene\n",
      "NLGN4Y  2.042237e-83  1.6773833 0.256 0.100 7.474790e-79  NLGN4Y\n",
      "IFI44L  2.897056e-42  1.3472702 0.304 0.184 1.060351e-37  IFI44L\n",
      "USP9Y   4.844771e-79  1.2559787 0.305 0.137 1.773235e-74   USP9Y\n",
      "UTY     9.179852e-95  1.0046086 0.418 0.199 3.359918e-90     UTY\n",
      "TMEM163 6.189402e-43  0.9244404 0.471 0.332 2.265383e-38 TMEM163\n",
      "SPARCL1 3.906306e-24  0.7416695 0.266 0.173 1.429747e-19 SPARCL1\n",
      "EPSTI1  1.635478e-29  0.7055519 0.412 0.292 5.986014e-25  EPSTI1\n",
      "DLEU7   7.554619e-27  0.6720786 0.344 0.237 2.765066e-22   DLEU7\n",
      "DDX60L  6.291981e-19  0.6648832 0.350 0.261 2.302928e-14  DDX60L\n",
      "XAF1    1.219088e-15  0.6404600 0.268 0.194 4.461983e-11    XAF1\n",
      "\n",
      "Top 10/20 genes for cell type: endothelia cells \n",
      "                 p_val avg_log2FC pct.1 pct.2    p_val_adj      gene\n",
      "IFIT1     1.734161e-14   2.000863 0.197 0.084 6.347204e-10     IFIT1\n",
      "NLGN4Y    1.988386e-21   1.876730 0.224 0.077 7.277692e-17    NLGN4Y\n",
      "OAS2      7.632820e-09   1.653122 0.159 0.079 2.793689e-04      OAS2\n",
      "IFIT2     2.064610e-07   1.631332 0.106 0.046 7.556680e-03     IFIT2\n",
      "ZFY       7.465924e-15   1.610245 0.172 0.064 2.732603e-10       ZFY\n",
      "OVCH1-AS1 8.123986e-07   1.600754 0.095 0.042 2.973460e-02 OVCH1-AS1\n",
      "IFIT3     5.302216e-08   1.445022 0.171 0.093 1.940664e-03     IFIT3\n",
      "MX2       6.000968e-07   1.410459 0.145 0.077 2.196414e-02       MX2\n",
      "USP9Y     2.275298e-23   1.346714 0.321 0.137 8.327819e-19     USP9Y\n",
      "MX1       8.315178e-08   1.286545 0.292 0.203 3.043438e-03       MX1\n"
     ]
    }
   ],
   "source": [
    "top10_tables <- list()\n",
    "\n",
    "for (cell_type in names(deg_results_by_cluster)) {\n",
    "  cat(\"\\nTop 10/20 genes for cell type:\", cell_type, \"\\n\")\n",
    "  \n",
    "  # Get the dataframe\n",
    "  df <- deg_results_by_cluster[[cell_type]] %>% filter(p_val_adj < 0.05)\n",
    "    \n",
    "  # Add gene names as a column if they're rownames\n",
    "  df$gene <- rownames(df)\n",
    "    \n",
    "  df <- df[!grepl(\"^(LINC)\", df$gene), ]\n",
    "  df <- df[!grepl(\"^(NPIPA8|TTTY)\", df$gene), ]\n",
    "  df <- df[!grepl(\"^(MT-)\", df$gene), ]\n",
    "  df <- df[!grepl(\"\\\\.\\\\d+$\", df$gene), ]\n",
    "    \n",
    "  if (cell_type %in% c(\"microglia\", \"inhibitory neurons\", \"astrocytes\")){\n",
    "      df <- df %>% filter(pct.1 >= 0.25 | pct.2 >= 0.25)\n",
    "  }\n",
    "  \n",
    "  # Select top 10 genes with highest avg_log2FC\n",
    "  top_genes <- df[order(-df$avg_log2FC), ][1:10, ]\n",
    "\n",
    "  if (cell_type %in% c(\"inhibitory neurons\", \"excitatory neurons\")){\n",
    "     top_genes <- df[order(-df$avg_log2FC), ][1:20, ]\n",
    "  }\n",
    "\n",
    "  # Save to list\n",
    "  top10_tables[[cell_type]] <- top_genes\n",
    "  \n",
    "  # Optionally print the top genes\n",
    "  print(top_genes)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e9a2d569-1da2-4c28-ab52-6d589f9d6dd7",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "“Scaling data with a low number of groups may produce misleading results”\n",
      "\u001b[1m\u001b[22mScale for \u001b[32mcolour\u001b[39m is already present.\n",
      "Adding another scale for \u001b[32mcolour\u001b[39m, which will replace the existing scale.\n",
      "\u001b[1m\u001b[22mScale for \u001b[32msize\u001b[39m is already present.\n",
      "Adding another scale for \u001b[32msize\u001b[39m, which will replace the existing scale.\n",
      "Warning message:\n",
      "“Scaling data with a low number of groups may produce misleading results”\n",
      "\u001b[1m\u001b[22mScale for \u001b[32mcolour\u001b[39m is already present.\n",
      "Adding another scale for \u001b[32mcolour\u001b[39m, which will replace the existing scale.\n",
      "\u001b[1m\u001b[22mScale for \u001b[32msize\u001b[39m is already present.\n",
      "Adding another scale for \u001b[32msize\u001b[39m, which will replace the existing scale.\n",
      "Warning message:\n",
      "“Scaling data with a low number of groups may produce misleading results”\n",
      "\u001b[1m\u001b[22mScale for \u001b[32mcolour\u001b[39m is already present.\n",
      "Adding another scale for \u001b[32mcolour\u001b[39m, which will replace the existing scale.\n",
      "\u001b[1m\u001b[22mScale for \u001b[32msize\u001b[39m is already present.\n",
      "Adding another scale for \u001b[32msize\u001b[39m, which will replace the existing scale.\n",
      "Warning message:\n",
      "“Scaling data with a low number of groups may produce misleading results”\n",
      "\u001b[1m\u001b[22mScale for \u001b[32mcolour\u001b[39m is already present.\n",
      "Adding another scale for \u001b[32mcolour\u001b[39m, which will replace the existing scale.\n",
      "\u001b[1m\u001b[22mScale for \u001b[32msize\u001b[39m is already present.\n",
      "Adding another scale for \u001b[32msize\u001b[39m, which will replace the existing scale.\n",
      "Warning message:\n",
      "“Scaling data with a low number of groups may produce misleading results”\n",
      "\u001b[1m\u001b[22mScale for \u001b[32mcolour\u001b[39m is already present.\n",
      "Adding another scale for \u001b[32mcolour\u001b[39m, which will replace the existing scale.\n",
      "\u001b[1m\u001b[22mScale for \u001b[32msize\u001b[39m is already present.\n",
      "Adding another scale for \u001b[32msize\u001b[39m, which will replace the existing scale.\n",
      "Warning message:\n",
      "“Scaling data with a low number of groups may produce misleading results”\n",
      "\u001b[1m\u001b[22mScale for \u001b[32mcolour\u001b[39m is already present.\n",
      "Adding another scale for \u001b[32mcolour\u001b[39m, which will replace the existing scale.\n",
      "\u001b[1m\u001b[22mScale for \u001b[32msize\u001b[39m is already present.\n",
      "Adding another scale for \u001b[32msize\u001b[39m, which will replace the existing scale.\n",
      "Warning message:\n",
      "“Scaling data with a low number of groups may produce misleading results”\n",
      "\u001b[1m\u001b[22mScale for \u001b[32mcolour\u001b[39m is already present.\n",
      "Adding another scale for \u001b[32mcolour\u001b[39m, which will replace the existing scale.\n",
      "\u001b[1m\u001b[22mScale for \u001b[32msize\u001b[39m is already present.\n",
      "Adding another scale for \u001b[32msize\u001b[39m, which will replace the existing scale.\n"
     ]
    }
   ],
   "source": [
    "custom_palette <- c('#2166ac','#92c5de','#f7f7f7','#f4a582','#b2182b')\n",
    "\n",
    "# Loop through each cell type in the DEG results\n",
    "for (cell_type in names(deg_results_by_cluster)) {\n",
    "  top_genes <- top10_tables[[cell_type]][1:10,\"gene\"]\n",
    "  if (cell_type %in% c(\"inhibitory neurons\", \"excitatory neurons\")){\n",
    "     top_genes <- top10_tables[[cell_type]][1:20,\"gene\"]\n",
    "  }\n",
    "  \n",
    "  # Subset Seurat object for this cell type\n",
    "  sub_obj <- subset(ifnb_testct, subset = celltype == cell_type)\n",
    "  \n",
    "  # Keep only genes that exist in this object\n",
    "  available_genes <- intersect(top_genes, rownames(sub_obj))\n",
    "  \n",
    "  # Skip if no genes available\n",
    "  if (length(available_genes) == 0) next\n",
    "\n",
    "  # Get plot data first\n",
    "  plot_obj <- DotPlot(\n",
    "    object   = sub_obj,\n",
    "    features = available_genes,\n",
    "    group.by = \"disease\"\n",
    "  )\n",
    "  plot_data <- plot_obj$data\n",
    "  \n",
    "  # Calculate dynamic size breaks for percent expressed\n",
    "  min_pct <- min(plot_data$pct.exp)\n",
    "  max_pct <- max(plot_data$pct.exp)\n",
    "  breaks <- seq(min_pct, max_pct, length.out = 5)\n",
    "  labels <- paste0(round(breaks), \"%\")\n",
    "  \n",
    "  # Now plot with dynamic size scale\n",
    "  p <- plot_obj +\n",
    "    scale_color_gradientn(colors = custom_palette) +\n",
    "    scale_size_area(\n",
    "      max_size = 15,\n",
    "      breaks = breaks,\n",
    "      labels = labels,\n",
    "      name = \"Percent Expressed\"\n",
    "    ) +\n",
    "    ggtitle(paste(\"Cell Type:\", cell_type, \"- Top 10/20 DEGs for PD vs HC\")) +\n",
    "    RotatedAxis() +\n",
    "    theme(axis.text.x = element_text(angle = 45, hjust = 1))\n",
    "  \n",
    "  # Save\n",
    "  out_file <- paste0(\"../../man_figs_pdf/deg_dot_plot/PD_vs_HC_CellType_\", cell_type, \".pdf\")\n",
    "  ggsave(filename = out_file, plot = p, device = \"pdf\", width = 10, height = 5)\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "525ca199",
   "metadata": {},
   "source": [
    "### DEG counts per cell type - Heatmap"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3ff5c1e8-5847-4f73-94d8-5e00a9d4681d",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Saved: ../man_figs_pdf/allct_top_deg_heatmap_PD_vs_HC.pdf\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# --- INPUTS ---------------------------------------------------------------\n",
    "xlsx_path <- \"../../DEG_GO_Analysis/DEGs_PD_vs_HC/Significant_DEGs_PD_vs_HC_ct.xlsx\"\n",
    "out_pdf   <- \"../../man_figs_pdf/allct_top_deg_heatmap_PD_vs_HC.pdf\"\n",
    "\n",
    "row_order_words <- c(\n",
    "  \"astrocytes\",\n",
    "  \"endothelia cells\",\n",
    "  \"excitatory neurons\",\n",
    "  \"inhibitory neurons\",\n",
    "  \"microglia\",\n",
    "  \"oligodendrocytes\",\n",
    "  \"oligodendrocyte precursor cells\"\n",
    ")\n",
    "\n",
    "abbr_map <- c(\n",
    "  \"excitatory neurons\"               = \"Exc\",\n",
    "  \"inhibitory neurons\"               = \"Inh\",\n",
    "  \"oligodendrocytes\"                 = \"Oli\",\n",
    "  \"oligodendrocyte precursor cells\"  = \"OPC\",\n",
    "  \"endothelia cells\"                 = \"Endo\",\n",
    "  \"endothelial cells\"                = \"Endo\",  # alias just in case\n",
    "  \"astrocytes\"                       = \"Ast\",\n",
    "  \"microglia\"                        = \"Mic\"\n",
    ")\n",
    "\n",
    "top_n_per_group <- 10  # << top 10 per cell type\n",
    "\n",
    "# --- 1) COLLECT THE UNION OF TOP GENES (ORDERED BY row_order_words) ---------\n",
    "stopifnot(exists(\"top10_tables\"), length(top10_tables) > 0)\n",
    "tt_norm <- setNames(top10_tables, tolower(names(top10_tables)))\n",
    "\n",
    "grouped_genes <- list()\n",
    "for (ct in row_order_words) {\n",
    "  key <- tolower(ct)\n",
    "  if (!key %in% names(tt_norm)) next\n",
    "  df <- tt_norm[[key]]\n",
    "  genes_ct <- if (\"gene\" %in% names(df)) df$gene else rownames(df)\n",
    "  genes_ct <- genes_ct[!is.na(genes_ct)]\n",
    "  genes_ct <- unique(genes_ct)\n",
    "  grouped_genes[[ct]] <- head(genes_ct, top_n_per_group)  # enforce top 10 per group\n",
    "}\n",
    "\n",
    "gene_order <- unlist(grouped_genes, use.names = FALSE)\n",
    "stopifnot(length(gene_order) > 0)\n",
    "\n",
    "# vertical separators between gene groups\n",
    "group_sizes <- vapply(grouped_genes, length, integer(1))\n",
    "vlines_x <- cumsum(group_sizes)\n",
    "if (length(vlines_x) > 0) vlines_x <- vlines_x[-length(vlines_x)] + 0.5\n",
    "\n",
    "# unique internal keys for columns; pretty labels can repeat\n",
    "gene_label     <- gene_order\n",
    "gene_key_order <- make.unique(gene_order, sep = \"_dup\")\n",
    "names(gene_label) <- gene_key_order  # key -> label\n",
    "\n",
    "# --- 2) READ LOG2FC FROM WORKBOOK ------------------------------------------\n",
    "sheet_names <- getSheetNames(xlsx_path)\n",
    "\n",
    "read_fc_vec <- function(sheet_nm) {\n",
    "  df <- read.xlsx(xlsx_path, sheet = sheet_nm, rowNames = TRUE)\n",
    "  if (nrow(df) == 0) return(setNames(numeric(0), character(0)))\n",
    "  nm <- tolower(names(df))\n",
    "  fc_col <- names(df)[match(TRUE, nm %in% c(\"avg_log2fc\",\"avg_log2_fc\",\"avg_logfc\",\"log2fc\"))]\n",
    "  if (is.na(fc_col)) stop(\"avg_log2FC column not found in sheet: \", sheet_nm)\n",
    "  g <- if (\"gene\" %in% names(df)) df$gene else rownames(df)\n",
    "  setNames(as.numeric(df[[fc_col]]), g)\n",
    "}\n",
    "\n",
    "sheet_for <- function(ct_word) {\n",
    "  idx <- which(tolower(sheet_names) == tolower(ct_word))\n",
    "  if (length(idx) == 0 && tolower(ct_word) == \"endothelia cells\") {\n",
    "    idx <- which(tolower(sheet_names) == \"endothelial cells\")\n",
    "  }\n",
    "  if (length(idx) == 0) return(NA_character_)\n",
    "  sheet_names[idx[1]]\n",
    "}\n",
    "\n",
    "row_sheets <- vapply(row_order_words, sheet_for, character(1))\n",
    "missing_rows <- is.na(row_sheets)\n",
    "if (any(missing_rows)) {\n",
    "  message(\"Warning: missing sheets for rows: \",\n",
    "          paste(row_order_words[missing_rows], collapse = \", \"))\n",
    "  row_order_words <- row_order_words[!missing_rows]\n",
    "  row_sheets      <- row_sheets[!missing_rows]\n",
    "}\n",
    "\n",
    "# build matrix (rows = cell types; cols = gene_key_order)\n",
    "mat <- matrix(0, nrow = length(row_order_words), ncol = length(gene_key_order))\n",
    "rownames(mat) <- row_order_words\n",
    "colnames(mat) <- gene_key_order\n",
    "\n",
    "# fill by aligning workbook vectors to gene_label (handles duplicates correctly)\n",
    "for (i in seq_along(row_sheets)) {\n",
    "  vec <- read_fc_vec(row_sheets[i])\n",
    "  vec <- vec[!is.na(names(vec))]\n",
    "  aligned <- vec[unname(gene_label)]      # repeats values for duplicated labels\n",
    "  aligned[is.na(aligned)] <- 0\n",
    "  mat[i, ] <- aligned\n",
    "}\n",
    "\n",
    "# y-axis abbreviations\n",
    "row_labels <- unname(abbr_map[tolower(rownames(mat))])\n",
    "row_labels[is.na(row_labels)] <- rownames(mat)[is.na(row_labels)]\n",
    "\n",
    "# --- 3) LONG DATA + COLOR LIMITS --------------------------------------------\n",
    "df_long <- as.data.frame(mat)\n",
    "df_long$celltype <- factor(row_labels, levels = row_labels)\n",
    "\n",
    "df_long <- tidyr::pivot_longer(\n",
    "  df_long,\n",
    "  cols = all_of(colnames(mat)),\n",
    "  names_to = \"gene_key\",\n",
    "  values_to = \"log2fc\"\n",
    ")\n",
    "\n",
    "all_vals <- df_long$log2fc[is.finite(df_long$log2fc)]\n",
    "lo <- stats::quantile(all_vals, 0.02, na.rm = TRUE)\n",
    "hi <- stats::quantile(all_vals, 0.98, na.rm = TRUE)\n",
    "lim_low  <- min(lo, -0.5)\n",
    "lim_high <- max(hi,  0.5)\n",
    "\n",
    "lab_map <- setNames(unname(gene_label), names(gene_label))  # key -> label\n",
    "\n",
    "# --- COLOR PALETTES ----------------------------------------------------------\n",
    "pal_down <- c('#f7fbff','#deebf7','#c6dbef','#9ecae1','#6baed6','#4292c6','#2171b5','#08519c','#08306b')\n",
    "pal_up   <- c('#fff5f0','#fee0d2','#fcbba1','#fc9272','#fb6a4a','#ef3b2c','#cb181d','#a50f15','#67000d')\n",
    "pal_all  <- c(rev(pal_down), \"#ffffff\", pal_up)\n",
    "\n",
    "p0  <- (0 - lim_low) / (lim_high - lim_low)\n",
    "eps <- 1e-6\n",
    "vals <- c(\n",
    "  seq(0, max(0, p0 - eps), length.out = length(pal_down)),\n",
    "  p0,\n",
    "  seq(min(1, p0 + eps), 1, length.out = length(pal_up))\n",
    ")\n",
    "\n",
    "# --- 4) PLOT -----------------------------------------------------------------\n",
    "p <- ggplot(df_long, aes(x = gene_key, y = celltype, fill = log2fc)) +\n",
    "  geom_tile(color = \"black\", linewidth = 0.25) +\n",
    "  scale_fill_gradientn(colors = pal_all, values = vals,\n",
    "                       limits = c(lim_low, lim_high),\n",
    "                       name = expression(log[2]*\"FC\"),\n",
    "                       na.value = \"#ffffff\") +\n",
    "  scale_x_discrete(position = \"top\",\n",
    "                   limits = gene_key_order,\n",
    "                   labels = lab_map[gene_key_order]) +\n",
    "  scale_y_discrete(position = \"left\", limits = rev(row_labels)) +\n",
    "  labs(x = NULL, y = NULL) +\n",
    "  theme_minimal(base_size = 12) +\n",
    "  theme(\n",
    "    panel.grid = element_blank(),\n",
    "    axis.text.y = element_text(face = \"bold\"),\n",
    "    axis.text.x = element_text(angle = 90, hjust = 0, vjust = 0.5, face = \"italic\"),\n",
    "    legend.title = element_text(face = \"bold\")\n",
    "  )\n",
    "\n",
    "if (length(vlines_x) > 0) {\n",
    "  p <- p + geom_vline(xintercept = vlines_x, color = \"black\", linewidth = 0.9)\n",
    "}\n",
    "\n",
    "w <- max(6, 0.25 * length(gene_key_order) + 2)\n",
    "h <- 2.2 + 0.45 * length(row_order_words)\n",
    "\n",
    "ggsave(out_pdf, plot = p, width = w, height = h, units = \"in\", device = \"pdf\")\n",
    "message(\"Saved: \", out_pdf)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5f1ecee5-2917-42f6-9b80-52cc929f6d09",
   "metadata": {},
   "source": [
    "## GO Enrichment analysis on DEGs for each cell type"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "b5ac6ee0-6243-4f12-8b89-47fbe4395326",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "go_analysis <- function(deg_results_by_cluster, log2FC_threshold = 0.0){\n",
    "    set.seed(384)\n",
    "    # Define the database(s) to use for EnrichR, focusing on GO Biological Process\n",
    "    dbs <- c(\"GO_Biological_Process_2021\")\n",
    "    \n",
    "    # Initialize a list to store enrichment results\n",
    "    enrichment_results <- list()\n",
    "    marker_counts <- list() \n",
    "    \n",
    "    # Loop over each cluster in deg_results_by_cluster to perform enrichment analysis\n",
    "    for (cluster_name in names(deg_results_by_cluster)) {\n",
    "      # Retrieve DEGs for the current cluster from deg_results_by_cluster\n",
    "      cluster_data <- deg_results_by_cluster[[cluster_name]]\n",
    "      \n",
    "      # Select significant DEGs (e.g., adjusted p-value < 0.05 and log fold change threshold)\n",
    "      significant_genes <- rownames(cluster_data[cluster_data$p_val_adj < 0.05 & cluster_data$avg_log2FC > log2FC_threshold, ])\n",
    "      marker_counts[[cluster_name]] <- length(significant_genes)\n",
    "        \n",
    "      # Perform enrichment analysis if there are significant genes\n",
    "      if (length(significant_genes) > 0) {\n",
    "        enrich_result <- enrichr(significant_genes, dbs)\n",
    "        go_results <- enrich_result[[\"GO_Biological_Process_2021\"]]\n",
    "        \n",
    "        # Sort results by adjusted p-value and select the top 10 processes\n",
    "        go_results <- go_results[order(go_results$Adjusted.P.value), ]\n",
    "        top_processes <- head(go_results, 15)\n",
    "    \n",
    "        if (nrow(top_processes) != 0) {\n",
    "            # Initialize list to store the average fold change for each GO term\n",
    "            avg_fc_for_terms <- numeric(nrow(top_processes))\n",
    "            \n",
    "            # Calculate the average fold change for each GO term in top_processes\n",
    "            for (j in 1:nrow(top_processes)) {\n",
    "              if (!is.na(top_processes$Genes[j]) && top_processes$Genes[j] != \"\") {\n",
    "                genes_in_term <- strsplit(as.character(top_processes$Genes[j]), \";\")[[1]]  # Split the genes string by semicolon\n",
    "                \n",
    "                # Only keep genes that are present in cluster_data\n",
    "                genes_in_term <- genes_in_term[genes_in_term %in% rownames(cluster_data)]\n",
    "                  \n",
    "                # Retrieve fold changes for these genes\n",
    "                if (length(genes_in_term) > 0) {\n",
    "                  fc_values <- cluster_data[genes_in_term, \"avg_log2FC\"]\n",
    "                  avg_fc_for_terms[j] <- mean(fc_values, na.rm = TRUE)  # Calculate the average, ignoring NA values\n",
    "                } else {\n",
    "                  avg_fc_for_terms[j] <- NA  # If no genes are found, set to NA\n",
    "                }\n",
    "              } else {\n",
    "                avg_fc_for_terms[j] <- NA  # If 'Genes' is NA, set to NA\n",
    "              }\n",
    "            }\n",
    "        \n",
    "            # Add the average fold changes as a new column in top_processes\n",
    "            top_processes$Avg_Log2FC <- avg_fc_for_terms\n",
    "        }\n",
    "        # Store the top GO processes for this cluster\n",
    "        enrichment_results[[cluster_name]] <- top_processes\n",
    "      }\n",
    "    }\n",
    "    return(list(marker_counts = marker_counts, enrichment_results = enrichment_results))\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "1ce9d041-7bc2-44d0-9604-4f7b6b49bff2",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Uploading data to Enrichr... Done.\n",
      "  Querying GO_Biological_Process_2021... Done.\n",
      "Parsing results... Done.\n",
      "Uploading data to Enrichr... Done.\n",
      "  Querying GO_Biological_Process_2021... Done.\n",
      "Parsing results... Done.\n",
      "Uploading data to Enrichr... Done.\n",
      "  Querying GO_Biological_Process_2021... Done.\n",
      "Parsing results... Done.\n",
      "Uploading data to Enrichr... Done.\n",
      "  Querying GO_Biological_Process_2021... Done.\n",
      "Parsing results... Done.\n",
      "Uploading data to Enrichr... Done.\n",
      "  Querying GO_Biological_Process_2021... Done.\n",
      "Parsing results... Done.\n",
      "Uploading data to Enrichr... Done.\n",
      "  Querying GO_Biological_Process_2021... Done.\n",
      "Parsing results... Done.\n",
      "Uploading data to Enrichr... Done.\n",
      "  Querying GO_Biological_Process_2021... Done.\n",
      "Parsing results... Done.\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "<environment: R_GlobalEnv>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "set.seed(384)\n",
    "deg_results_by_cluster_filt <- lapply(deg_results_by_cluster, function(df) {\n",
    "  genes <- rownames(df)\n",
    "  keep  <- (!grepl(\"^(RPL|RPS)\", genes)) & (!grepl(\"^MT-\", genes)) & (!grepl(\"\\\\.\", genes))\n",
    "  df[keep, , drop = FALSE]\n",
    "})\n",
    "output <- go_analysis(deg_results_by_cluster_filt, log2FC_threshold = 0.25)\n",
    "list2env(output, envir = .GlobalEnv)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "a27fdcb4-0c0b-43e1-9352-37cfbc5579b7",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<dl>\n",
       "\t<dt>$`excitatory neurons`</dt>\n",
       "\t\t<dd>281</dd>\n",
       "\t<dt>$`oligodendrocyte precursor cells`</dt>\n",
       "\t\t<dd>159</dd>\n",
       "\t<dt>$oligodendrocytes</dt>\n",
       "\t\t<dd>889</dd>\n",
       "\t<dt>$`inhibitory neurons`</dt>\n",
       "\t\t<dd>330</dd>\n",
       "\t<dt>$astrocytes</dt>\n",
       "\t\t<dd>1557</dd>\n",
       "\t<dt>$microglia</dt>\n",
       "\t\t<dd>400</dd>\n",
       "\t<dt>$`endothelia cells`</dt>\n",
       "\t\t<dd>29</dd>\n",
       "</dl>\n"
      ],
      "text/latex": [
       "\\begin{description}\n",
       "\\item[\\$`excitatory neurons`] 281\n",
       "\\item[\\$`oligodendrocyte precursor cells`] 159\n",
       "\\item[\\$oligodendrocytes] 889\n",
       "\\item[\\$`inhibitory neurons`] 330\n",
       "\\item[\\$astrocytes] 1557\n",
       "\\item[\\$microglia] 400\n",
       "\\item[\\$`endothelia cells`] 29\n",
       "\\end{description}\n"
      ],
      "text/markdown": [
       "$`excitatory neurons`\n",
       ":   281\n",
       "$`oligodendrocyte precursor cells`\n",
       ":   159\n",
       "$oligodendrocytes\n",
       ":   889\n",
       "$`inhibitory neurons`\n",
       ":   330\n",
       "$astrocytes\n",
       ":   1557\n",
       "$microglia\n",
       ":   400\n",
       "$`endothelia cells`\n",
       ":   29\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "$`excitatory neurons`\n",
       "[1] 281\n",
       "\n",
       "$`oligodendrocyte precursor cells`\n",
       "[1] 159\n",
       "\n",
       "$oligodendrocytes\n",
       "[1] 889\n",
       "\n",
       "$`inhibitory neurons`\n",
       "[1] 330\n",
       "\n",
       "$astrocytes\n",
       "[1] 1557\n",
       "\n",
       "$microglia\n",
       "[1] 400\n",
       "\n",
       "$`endothelia cells`\n",
       "[1] 29\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "marker_counts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b041da4a-4d12-4e1c-8c31-e2509ea58bdc",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Logp Plot saved\"\n",
      "[1] \"Log2FC File saved\"\n"
     ]
    }
   ],
   "source": [
    "set.seed(384) \n",
    "# Prepare matrices for plotting\n",
    "cluster_names <- c(\"astrocytes\", \"endothelia cells\", \"excitatory neurons\", \"inhibitory neurons\", \"microglia\", \"oligodendrocytes\", \"oligodendrocyte precursor cells\")\n",
    "# Collect unique biological processes across all clusters\n",
    "all_processes <- unique(unlist(lapply(enrichment_results, function(x) x$Term)))\n",
    "enrichment_matrix <- matrix(\n",
    "  NA, nrow = length(all_processes), ncol = length(cluster_names),\n",
    "  dimnames = list(all_processes, cluster_names)\n",
    ")\n",
    "fc_matrix <- matrix(\n",
    "  NA, nrow = length(all_processes), ncol = length(cluster_names),\n",
    "  dimnames = list(all_processes, cluster_names)\n",
    ")\n",
    "\n",
    "# Fill the matrices with -log10(adjusted p-value) and Avg_Log2FC for each process and cluster\n",
    "for (cluster_name in cluster_names) {\n",
    "  if (!is.null(enrichment_results[[cluster_name]]) && nrow(enrichment_results[[cluster_name]]) > 0) {\n",
    "    for (j in 1:nrow(enrichment_results[[cluster_name]])) {\n",
    "      process <- enrichment_results[[cluster_name]]$Term[j]\n",
    "      pval <- enrichment_results[[cluster_name]]$Adjusted.P.value[j]\n",
    "      avg_log2fc <- enrichment_results[[cluster_name]]$Avg_Log2FC[j]\n",
    "  \n",
    "      # Fill p-value matrix with -log10(p-value)\n",
    "      enrichment_matrix[process, cluster_name] <- -log10(pval)\n",
    "      \n",
    "      # Fill fold change matrix with Avg_Log2FC\n",
    "      fc_matrix[process, cluster_name] <- avg_log2fc\n",
    "      \n",
    "    }\n",
    "  }\n",
    "}   \n",
    "\n",
    "# Replace NA values with a low number (e.g., 0) for better plotting\n",
    "enrichment_matrix[is.na(enrichment_matrix)] <- 0\n",
    "\n",
    "output_pdf = \"../../man_figs_pdf/allct_deg_go_PDvsHC_logp+.pdf\"\n",
    "# Plot the heatmap using pheatmap\n",
    "pheatmap::pheatmap(enrichment_matrix,\n",
    "                   color = c(\n",
    "                     colorRampPalette(c(\"white\", '#ef3b2c'))(10),  # More resolution for small values\n",
    "                     colorRampPalette(c('#ef3b2c','#cb181d','#a50f15','#67000d'))(90) # High contrast for larger values\n",
    "                   ),\n",
    "                   na_col = \"grey\",           # Use grey for NA values (e.g., processes not enriched in clusters)\n",
    "                   main = \"Enriched BP of Microglial State Signatures (-log10 p-value)\",\n",
    "                   cluster_rows = FALSE,       # Cluster processes for better visualization\n",
    "                   cluster_cols = FALSE,      # Do not cluster clusters (retain the original order)\n",
    "                   angle_col = 45,            # Rotate column names for readability\n",
    "                   fontsize_row = 6,          # Font size for process names\n",
    "                   fontsize_col = 8,\n",
    "                   filename = output_pdf)         # Font size for cluster labels\n",
    "print(\"Logp Plot saved\")\n",
    "\n",
    "# Replace NA values with a low number (e.g., 0) for better plotting\n",
    "fc_matrix[is.na(fc_matrix)] <- 0\n",
    "\n",
    "# Define the custom color palettes\n",
    "# negative_colors <- colorRampPalette(c('#f7fbff','#deebf7','#c6dbef','#9ecae1','#6baed6','#4292c6', '#2171b5', '#084594'))(50)\n",
    "negative_colors <- colorRampPalette(c('#084594','#2171b5','#4292c6','#6baed6','#9ecae1','#c6dbef', '#deebf7', '#f7fbff'))(50)\n",
    "positive_colors <- colorRampPalette(c('#fff5f0','#fee0d2','#fcbba1','#fc9272','#fb6a4a','#ef3b2c','#cb181d','#99000d'))(50)\n",
    "\n",
    "# Combine palettes with white at 0\n",
    "custom_colors <- c(negative_colors, \"white\", positive_colors)\n",
    "\n",
    "# Define custom breaks to center the color scale at 0\n",
    "max_abs_value <- max(abs(fc_matrix))\n",
    "breaks <- seq(-max_abs_value, max_abs_value, length.out = length(custom_colors) + 1)\n",
    "\n",
    "output_pdf = \"../../man_figs_pdf/allct_deg_go_PDvsHC_log2fc+.pdf\"\n",
    "pheatmap::pheatmap(fc_matrix,\n",
    "                   color = custom_colors,\n",
    "                   breaks = breaks,              # Use custom breaks for diverging colors\n",
    "                   na_col = \"grey\",              # Use grey for NA values\n",
    "                   main = \"Average Log2 Fold Change\",\n",
    "                   cluster_rows = FALSE,          # Cluster processes for better visualization\n",
    "                   cluster_cols = FALSE,          # Cluster clusters for better visualization\n",
    "                   angle_col = 45,               # Rotate column names for readability\n",
    "                   fontsize_row = 6,             # Font size for process names\n",
    "                   fontsize_col = 8,\n",
    "                   filename = output_pdf)             # Font size for cluster labels\n",
    "print(\"Log2FC File saved\")                                                  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3ff017ac-e2ab-4fc5-90ba-9651d392f43d",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "library(openxlsx)\n",
    "# Create a new workbook\n",
    "wb <- createWorkbook()\n",
    "\n",
    "for (cluster_name in cluster_names) {\n",
    "    # Filter the data for each cluster\n",
    "    cluster_data <- enrichment_results[[cluster_name]]\n",
    "    \n",
    "    # Add a worksheet with the cluster name\n",
    "    addWorksheet(wb, sheetName = cluster_name)\n",
    "    \n",
    "    # Write the filtered data to the worksheet\n",
    "    writeData(wb, sheet = cluster_name, cluster_data, rowNames = TRUE)\n",
    "}\n",
    "\n",
    "# Save the workbook to an Excel file\n",
    "saveWorkbook(wb, file = \"../../DEG_GO_Analysis/GO_PD_vs_HC/GO_Results_PD_vs_HC_allct+.xlsx\", overwrite = TRUE)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bdae2002-8d96-468e-830f-0e9c01fc6aa6",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "saveRDS(enrichment_results, \"../../DEG_GO_Analysis/GO_PD_vs_HC/cached_enrichment_results_allct+.rds\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2f278367-b136-41c4-bb84-4b3a5e6ecd6e",
   "metadata": {},
   "source": [
    "# DEG analysis (RBD vs HC) for each cell type"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "f6799066-69ec-4091-9cec-27a7510843fb",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "set.seed(384)\n",
    "Idents(ifnb_testct) <- \"celltype\"\n",
    "deg_results_by_cluster <- list()\n",
    "\n",
    "\n",
    "for (ct in unique(ifnb_testct$celltype)) {\n",
    "  ct_subset <- subset(ifnb_testct, idents = ct)\n",
    "  Idents(ct_subset) <- \"disease\"\n",
    "  \n",
    "  # Check if both \"PD\" and \"HC\" are present in the subset\n",
    "  if (all(c(\"RBD\", \"HC\") %in% unique(Idents(ct_subset)))) {\n",
    "    # Perform differential expression analysis between \"PD\" and \"HC\" within this cluster\n",
    "    deg_results <- FindMarkers(ct_subset, ident.1 = \"RBD\", ident.2 = \"HC\")\n",
    "    deg_results_by_cluster[[ct]] <- deg_results\n",
    "  } else {\n",
    "    message(paste(\"Skipping ct\", ct, \"- does not contain both RBD and HC cells\"))\n",
    "  }\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "97eda562-d44f-4eeb-88dd-7edae41d7fb5",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "library(openxlsx)\n",
    "\n",
    "# Create a new workbook\n",
    "wb <- createWorkbook()\n",
    "\n",
    "for (ct in unique(ifnb_testct$celltype)) {\n",
    "    # Filter the data for each cluster\n",
    "    cluster_data <- deg_results_by_cluster[[ct]] %>% filter(p_val_adj < 0.05) %>% arrange(desc(avg_log2FC))\n",
    "    \n",
    "    # Add a worksheet with the cluster name\n",
    "    addWorksheet(wb, sheetName = ct)\n",
    "    \n",
    "    # Write the filtered data to the worksheet\n",
    "    writeData(wb, sheet = ct, cluster_data, rowNames = TRUE)\n",
    "}\n",
    "\n",
    "# Save the workbook to an Excel file\n",
    "saveWorkbook(wb, file = \"../../DEG_GO_Analysis/DEGs_RBD_vs_HC/Significant_DEGs_RBD_vs_HC_ct.xlsx\", overwrite = TRUE)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f4014501-59c0-475c-b98b-4b8261513027",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'excitatory neurons'</li><li>'oligodendrocyte precursor cells'</li><li>'oligodendrocytes'</li><li>'inhibitory neurons'</li><li>'astrocytes'</li><li>'microglia'</li><li>'endothelia cells'</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 'excitatory neurons'\n",
       "\\item 'oligodendrocyte precursor cells'\n",
       "\\item 'oligodendrocytes'\n",
       "\\item 'inhibitory neurons'\n",
       "\\item 'astrocytes'\n",
       "\\item 'microglia'\n",
       "\\item 'endothelia cells'\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 'excitatory neurons'\n",
       "2. 'oligodendrocyte precursor cells'\n",
       "3. 'oligodendrocytes'\n",
       "4. 'inhibitory neurons'\n",
       "5. 'astrocytes'\n",
       "6. 'microglia'\n",
       "7. 'endothelia cells'\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[1] \"excitatory neurons\"              \"oligodendrocyte precursor cells\"\n",
       "[3] \"oligodendrocytes\"                \"inhibitory neurons\"             \n",
       "[5] \"astrocytes\"                      \"microglia\"                      \n",
       "[7] \"endothelia cells\"               "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "library(openxlsx)\n",
    "\n",
    "xlsx_path <- \"../../DEG_GO_Analysis/DEGs_RBD_vs_HC/Significant_DEGs_RBD_vs_HC_ct.xlsx\"\n",
    "\n",
    "# Get sheet names (these become the sublist names)\n",
    "sheets <- getSheetNames(xlsx_path)\n",
    "\n",
    "# Optionally drop any sheet you don't want\n",
    "sheets <- setdiff(sheets, \"pericytes\")\n",
    "\n",
    "# Read each sheet -> data.frame with gene rownames\n",
    "deg_results_by_cluster <- setNames(\n",
    "  lapply(sheets, function(sh) {\n",
    "    df <- read.xlsx(xlsx_path, sheet = sh, rowNames = TRUE)\n",
    "\n",
    "    # Make sure numeric columns are numeric (robust to minor name variants)\n",
    "    nm <- tolower(names(df))\n",
    "    num_cols <- names(df)[nm %in% c(\"p_val\",\"avg_log2fc\",\"avg_log2_fc\",\"avg_logfc\",\n",
    "                                    \"pct.1\",\"pct1\",\"pct.2\",\"pct2\",\"p_val_adj\",\"padj\",\"p_adj\")]\n",
    "    for (cc in num_cols) df[[cc]] <- as.numeric(df[[cc]])\n",
    "\n",
    "    df\n",
    "  }),\n",
    "  sheets\n",
    ")\n",
    "\n",
    "# quick checks\n",
    "names(deg_results_by_cluster)\n",
    "# e.g. deg_results_by_cluster[[\"astrocytes\"]][1:5, ]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9d17f17c",
   "metadata": {},
   "source": [
    "### DEG counts per cell type - Heatmap"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cf3f4873-5f13-4c29-995a-9b039b99310b",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "“package ‘ggnewscale’ was built under R version 4.3.3”\n"
     ]
    }
   ],
   "source": [
    "suppressPackageStartupMessages({\n",
    "  library(openxlsx)\n",
    "  library(dplyr)\n",
    "  library(tidyr)\n",
    "  library(ggplot2)\n",
    "  library(ggnewscale)  # install.packages(\"ggnewscale\") if needed\n",
    "})\n",
    "\n",
    "# ---- paths ----\n",
    "xlsx_path <- \"../../DEG_GO_Analysis/DEGs_RBD_vs_HC/Significant_DEGs_RBD_vs_HC_ct.xlsx\"\n",
    "out_pdf   <- \"../../man_figs_pdf/allct_DEG_counts_heatmap_RBD_vs_HC.pdf\"\n",
    "\n",
    "# ---- settings ----\n",
    "fc_thresh <- 0.25\n",
    "desired_order <- c(\n",
    "  \"astrocytes\",\n",
    "  \"endothelia cells\",\n",
    "  \"excitatory neurons\",\n",
    "  \"inhibitory neurons\",\n",
    "  \"microglia\",\n",
    "  \"oligodendrocytes\",\n",
    "  \"oligodendrocyte precursor cells\"\n",
    ")\n",
    "\n",
    "# ---- read sheets and count ----\n",
    "sheets <- setdiff(getSheetNames(xlsx_path), \"pericytes\")\n",
    "\n",
    "count_list <- lapply(sheets, function(ct) {\n",
    "  df <- read.xlsx(xlsx_path, sheet = ct, rowNames = TRUE)\n",
    "  if (nrow(df) == 0) return(tibble(celltype = ct, Up = 0L, Down = 0L))\n",
    "\n",
    "  nm <- tolower(names(df))\n",
    "  fc_col   <- names(df)[which(nm %in% c(\"avg_log2fc\",\"avg_log2_fc\",\"avg_logfc\",\"log2fc\"))[1]]\n",
    "  padj_col <- names(df)[which(nm %in% c(\"p_val_adj\",\"padj\",\"p_adj\",\"p.adjusted\"))[1]]\n",
    "  if (is.na(fc_col)) stop(\"Could not find log2FC column in sheet: \", ct)\n",
    "\n",
    "  fc <- as.numeric(df[[fc_col]])\n",
    "  keep <- is.finite(fc) & abs(fc) > fc_thresh\n",
    "  if (!is.na(padj_col)) keep <- keep & is.finite(df[[padj_col]]) & (as.numeric(df[[padj_col]]) < 0.05)\n",
    "\n",
    "  fc <- fc[keep]\n",
    "\n",
    "  tibble(\n",
    "    celltype = ct,\n",
    "    Up   = sum(fc > 0, na.rm = TRUE),\n",
    "    Down = sum(fc < 0, na.rm = TRUE)\n",
    "  )\n",
    "})\n",
    "\n",
    "counts <- bind_rows(count_list)\n",
    "\n",
    "# ---- order rows per requested biology order (case-insensitive) ----\n",
    "counts <- counts %>%\n",
    "  mutate(.order_idx = match(tolower(celltype), tolower(desired_order))) %>%\n",
    "  arrange(.order_idx, .by_group = FALSE) %>%\n",
    "  select(-.order_idx)\n",
    "\n",
    "# ---- long format ----\n",
    "heat_df <- counts %>%\n",
    "  mutate(celltype = factor(celltype, levels = rev(celltype))) %>%\n",
    "  pivot_longer(c(Down, Up), names_to = \"Direction\", values_to = \"Count\")\n",
    "\n",
    "# Palettes (independent scales for each column)\n",
    "pal_down <- c('#f7fbff','#deebf7','#c6dbef','#9ecae1','#6baed6','#4292c6','#2171b5','#08519c','#08306b')\n",
    "pal_up   <- c('#fff5f0','#fee0d2','#fcbba1','#fc9272','#fb6a4a','#ef3b2c','#cb181d','#a50f15','#67000d')\n",
    "\n",
    "max_down <- max(heat_df$Count[heat_df$Direction == \"Down\"], na.rm = TRUE)\n",
    "max_up   <- max(heat_df$Count[heat_df$Direction == \"Up\"],   na.rm = TRUE)\n",
    "\n",
    "# ---- plot ----\n",
    "p <- ggplot() +\n",
    "  geom_tile(\n",
    "    data = dplyr::filter(heat_df, Direction == \"Down\"),\n",
    "    aes(x = Direction, y = celltype, fill = Count),\n",
    "    color = \"black\", linewidth = 0.5\n",
    "  ) +\n",
    "  scale_fill_gradientn(colors = pal_down, limits = c(0, max_down)) +\n",
    "  ggnewscale::new_scale_fill() +\n",
    "  geom_tile(\n",
    "    data = dplyr::filter(heat_df, Direction == \"Up\"),\n",
    "    aes(x = Direction, y = celltype, fill = Count),\n",
    "    color = \"black\", linewidth = 0.5\n",
    "  ) +\n",
    "  scale_fill_gradientn(colors = pal_up, limits = c(0, max_up)) +\n",
    "  geom_text(data = heat_df, aes(x = Direction, y = celltype, label = Count), size = 3.6) +\n",
    "  scale_x_discrete(position = \"top\") +\n",
    "  labs(x = NULL, y = NULL) +\n",
    "  theme_minimal(base_size = 12) +\n",
    "  theme(\n",
    "    panel.grid = element_blank(),\n",
    "    legend.position = \"none\",\n",
    "    axis.text.x = element_text(face = \"bold\", margin = margin(b = 4)),\n",
    "    axis.text.y = element_text(face = \"bold\")\n",
    "  )\n",
    "\n",
    "ht <- max(3, 0.45 * length(levels(heat_df$celltype)) + 1)\n",
    "ggsave(out_pdf, plot = p, width = 3.2, height = ht, units = \"in\", device = \"pdf\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1a693ee0",
   "metadata": {},
   "source": [
    "### Dot plot for top DEGs per cell type"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "a9bf4e1e-5768-4b1e-b87d-94def4871e1a",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Top 10/20 genes for cell type: excitatory neurons \n",
      "                 p_val avg_log2FC pct.1 pct.2     p_val_adj     gene\n",
      "SERPINA3  2.499571e-37   3.189761 0.015 0.003  9.148681e-33 SERPINA3\n",
      "CHI3L1    1.842128e-82   2.917046 0.044 0.011  6.742373e-78   CHI3L1\n",
      "PRKY      0.000000e+00   2.667545 0.271 0.048  0.000000e+00     PRKY\n",
      "CD44      5.211069e-48   2.573488 0.029 0.008  1.907303e-43     CD44\n",
      "KIF25    1.085900e-111   2.381427 0.046 0.009 3.974503e-107    KIF25\n",
      "TBL1Y     1.735052e-21   2.150881 0.011 0.003  6.350462e-17    TBL1Y\n",
      "KDM5D     0.000000e+00   2.140323 0.542 0.125  0.000000e+00    KDM5D\n",
      "USP9Y     0.000000e+00   2.121901 0.930 0.210  0.000000e+00    USP9Y\n",
      "RPS4Y1    0.000000e+00   2.099902 0.270 0.065  0.000000e+00   RPS4Y1\n",
      "NLGN4Y    0.000000e+00   2.093563 0.905 0.207  0.000000e+00   NLGN4Y\n",
      "TMSB4Y    7.195475e-28   2.077732 0.013 0.003  2.633616e-23   TMSB4Y\n",
      "CD24     2.429056e-110   2.056773 0.054 0.013 8.890588e-106     CD24\n",
      "ZFY       0.000000e+00   2.002249 0.300 0.075  0.000000e+00      ZFY\n",
      "UTY       0.000000e+00   1.991299 0.894 0.206  0.000000e+00      UTY\n",
      "GRB7      1.331747e-15   1.962325 0.010 0.003  4.874325e-11     GRB7\n",
      "EIF1AY   3.213970e-121   1.940713 0.064 0.016 1.176345e-116   EIF1AY\n",
      "DDX3Y     0.000000e+00   1.882432 0.346 0.092  0.000000e+00    DDX3Y\n",
      "SLC26A3  4.911068e-150   1.796042 0.122 0.045 1.797500e-145  SLC26A3\n",
      "SLC22A18  2.059783e-47   1.778528 0.030 0.009  7.539010e-43 SLC22A18\n",
      "ZFY-AS1   1.963301e-30   1.760326 0.020 0.007  7.185880e-26  ZFY-AS1\n",
      "\n",
      "Top 10/20 genes for cell type: oligodendrocyte precursor cells \n",
      "                p_val avg_log2FC pct.1 pct.2    p_val_adj     gene\n",
      "FAM20A   2.602181e-07   2.521213 0.027 0.010 9.524243e-03   FAM20A\n",
      "CASZ1    2.613417e-19   2.294094 0.034 0.006 9.565369e-15    CASZ1\n",
      "CDCP1    7.488743e-11   2.216052 0.021 0.004 2.740955e-06    CDCP1\n",
      "CLMP     1.177531e-10   2.109391 0.027 0.007 4.309882e-06     CLMP\n",
      "RXFP1    1.188445e-10   2.086720 0.027 0.007 4.349829e-06    RXFP1\n",
      "TBL1Y    1.551122e-07   2.051657 0.014 0.003 5.677260e-03    TBL1Y\n",
      "COL6A3   1.252143e-06   2.012490 0.019 0.006 4.582969e-02   COL6A3\n",
      "SULF1    9.378202e-07   1.985212 0.026 0.010 3.432516e-02    SULF1\n",
      "KIAA0040 7.931121e-20   1.953088 0.065 0.021 2.902870e-15 KIAA0040\n",
      "RDH10    1.733391e-15   1.907992 0.075 0.032 6.344386e-11    RDH10\n",
      "\n",
      "Top 10/20 genes for cell type: oligodendrocytes \n",
      "                 p_val avg_log2FC pct.1 pct.2     p_val_adj     gene\n",
      "NPFFR2   1.691528e-158   5.047075 0.018 0.001 6.191160e-154   NPFFR2\n",
      "SPTSSB   4.385132e-232   4.913561 0.027 0.001 1.605002e-227   SPTSSB\n",
      "IL1RAPL2  0.000000e+00   4.869923 0.115 0.007  0.000000e+00 IL1RAPL2\n",
      "NRG1      0.000000e+00   4.849456 0.306 0.027  0.000000e+00     NRG1\n",
      "OCA2      0.000000e+00   4.840352 0.064 0.003  0.000000e+00     OCA2\n",
      "RXFP1     0.000000e+00   4.802023 0.105 0.005  0.000000e+00    RXFP1\n",
      "MCHR2     0.000000e+00   4.784683 0.073 0.004  0.000000e+00    MCHR2\n",
      "DPP4     4.549242e-224   4.777785 0.027 0.001 1.665068e-219     DPP4\n",
      "STAT4     0.000000e+00   4.669971 0.129 0.007  0.000000e+00    STAT4\n",
      "BTC      3.710739e-137   4.606809 0.016 0.001 1.358167e-132      BTC\n",
      "\n",
      "Top 10/20 genes for cell type: inhibitory neurons \n",
      "                p_val avg_log2FC pct.1 pct.2     p_val_adj    gene\n",
      "USP9Y    0.000000e+00  1.8742676 0.790 0.241  0.000000e+00   USP9Y\n",
      "KDM5D    0.000000e+00  1.8327476 0.355 0.118  0.000000e+00   KDM5D\n",
      "UTY      0.000000e+00  1.7152734 0.757 0.236  0.000000e+00     UTY\n",
      "NLGN4Y   0.000000e+00  1.6941195 0.749 0.237  0.000000e+00  NLGN4Y\n",
      "PCDH11Y  0.000000e+00  1.5981702 0.397 0.137  0.000000e+00 PCDH11Y\n",
      "ARL17B  6.691853e-154  1.0062543 0.379 0.233 2.449285e-149  ARL17B\n",
      "ST18     1.199794e-90  0.8896309 0.257 0.154  4.391365e-86    ST18\n",
      "PLCG2    0.000000e+00  0.8521909 0.796 0.461  0.000000e+00   PLCG2\n",
      "PLP1     1.036330e-52  0.7701910 0.282 0.202  3.793072e-48    PLP1\n",
      "PLEKHH1  1.810855e-25  0.7431931 0.269 0.216  6.627911e-21 PLEKHH1\n",
      "PTPRK    3.517611e-79  0.6992974 0.432 0.320  1.287481e-74   PTPRK\n",
      "MYO9B    5.791546e-37  0.6971368 0.256 0.192  2.119764e-32   MYO9B\n",
      "FGFR2    6.549567e-22  0.6829136 0.255 0.209  2.397207e-17   FGFR2\n",
      "ST6GAL1  7.575312e-15  0.6663748 0.250 0.215  2.772640e-10 ST6GAL1\n",
      "GRM3     9.801996e-27  0.6585943 0.271 0.219  3.587629e-22    GRM3\n",
      "TMEM144  3.632976e-31  0.6495960 0.252 0.195  1.329706e-26 TMEM144\n",
      "DPYD     3.239433e-52  0.6465776 0.410 0.327  1.185665e-47    DPYD\n",
      "IGSF9B   1.520535e-22  0.6362479 0.277 0.228  5.565310e-18  IGSF9B\n",
      "CERS4    3.326371e-55  0.6263779 0.334 0.247  1.217485e-50   CERS4\n",
      "FSTL4    4.589232e-25  0.6116359 0.285 0.233  1.679705e-20   FSTL4\n",
      "\n",
      "Top 10/20 genes for cell type: astrocytes \n",
      "                 p_val avg_log2FC pct.1 pct.2     p_val_adj     gene\n",
      "CHI3L1    0.000000e+00   2.950172 0.369 0.076  0.000000e+00   CHI3L1\n",
      "WARS      0.000000e+00   2.942554 0.268 0.072  0.000000e+00     WARS\n",
      "GALNT16   0.000000e+00   2.603265 0.288 0.058  0.000000e+00  GALNT16\n",
      "SERPINA3  0.000000e+00   2.518696 0.292 0.063  0.000000e+00 SERPINA3\n",
      "ARHGEF3   0.000000e+00   2.199892 0.604 0.234  0.000000e+00  ARHGEF3\n",
      "CD44      0.000000e+00   2.039261 0.479 0.183  0.000000e+00     CD44\n",
      "FAM20C    0.000000e+00   1.782755 0.303 0.103  0.000000e+00   FAM20C\n",
      "TPST1     0.000000e+00   1.764644 0.707 0.419  0.000000e+00    TPST1\n",
      "USP9Y     0.000000e+00   1.726964 0.563 0.169  0.000000e+00    USP9Y\n",
      "CRISPLD1 1.214883e-302   1.724620 0.287 0.099 4.446595e-298 CRISPLD1\n",
      "\n",
      "Top 10/20 genes for cell type: microglia \n",
      "                p_val avg_log2FC pct.1 pct.2     p_val_adj    gene\n",
      "TPRG1   3.689570e-176   2.941437 0.280 0.061 1.350419e-171   TPRG1\n",
      "PRKY    1.433157e-153   2.050634 0.274 0.065 5.245497e-149    PRKY\n",
      "NLGN4Y  5.531503e-142   1.856118 0.320 0.100 2.024585e-137  NLGN4Y\n",
      "FCGBP    5.577248e-70   1.776810 0.306 0.151  2.041328e-65   FCGBP\n",
      "PVT1    6.796378e-119   1.666947 0.342 0.131 2.487542e-114    PVT1\n",
      "USP9Y   1.574258e-194   1.648105 0.430 0.137 5.761941e-190   USP9Y\n",
      "BCL6    2.815809e-114   1.622390 0.361 0.151 1.030614e-109    BCL6\n",
      "TMEM163 2.667508e-164   1.387642 0.623 0.332 9.763348e-160 TMEM163\n",
      "TBC1D8  3.873403e-111   1.364239 0.439 0.214 1.417704e-106  TBC1D8\n",
      "CD163    1.266484e-96   1.247868 0.379 0.174  4.635459e-92   CD163\n",
      "\n",
      "Top 10/20 genes for cell type: endothelia cells \n",
      "                p_val avg_log2FC pct.1 pct.2    p_val_adj     gene\n",
      "SCT      3.345269e-12   5.472326 0.036 0.001 1.224402e-07      SCT\n",
      "HLA-DRB5 2.018540e-09   2.991384 0.042 0.006 7.388059e-05 HLA-DRB5\n",
      "STC1     6.678364e-20   2.987701 0.167 0.051 2.444348e-15     STC1\n",
      "TBL1Y    5.876655e-10   2.626530 0.055 0.011 2.150915e-05    TBL1Y\n",
      "ACKR1    1.289376e-08   2.431348 0.071 0.022 4.719244e-04    ACKR1\n",
      "GPRC5A   3.907285e-10   2.321010 0.051 0.008 1.430105e-05   GPRC5A\n",
      "MMRN1    5.187369e-07   2.207120 0.081 0.032 1.898629e-02    MMRN1\n",
      "PRKY     1.215975e-76   2.189362 0.476 0.120 4.450590e-72     PRKY\n",
      "SEMA3F   8.442950e-20   1.841559 0.192 0.065 3.090204e-15   SEMA3F\n",
      "EIF1AY   1.187895e-07   1.833126 0.075 0.027 4.347816e-03   EIF1AY\n"
     ]
    }
   ],
   "source": [
    "top10_tables <- list()\n",
    "\n",
    "for (cell_type in names(deg_results_by_cluster)) {\n",
    "  cat(\"\\nTop 10/20 genes for cell type:\", cell_type, \"\\n\")\n",
    "  \n",
    "  # Get the dataframe\n",
    "  df <- deg_results_by_cluster[[cell_type]] %>% filter(p_val_adj < 0.05)\n",
    "    \n",
    "  # Add gene names as a column if they're rownames\n",
    "  df$gene <- rownames(df)\n",
    "    \n",
    "  df <- df[!grepl(\"^(LINC)\", df$gene), ]\n",
    "  df <- df[!grepl(\"^(NPIPA8|TTTY)\", df$gene), ]\n",
    "  df <- df[!grepl(\"^(MT-)\", df$gene), ]\n",
    "  df <- df[!grepl(\"\\\\.\\\\d+$\", df$gene), ]\n",
    "    \n",
    "  if (cell_type %in% c(\"microglia\", \"inhibitory neurons\", \"astrocytes\")){\n",
    "      df <- df %>% filter(pct.1 >= 0.25 | pct.2 >= 0.25)\n",
    "  }\n",
    "  \n",
    "  # Select top 10 genes with highest avg_log2FC\n",
    "  top_genes <- df[order(-df$avg_log2FC), ][1:10, ]\n",
    "\n",
    "  if (cell_type %in% c(\"inhibitory neurons\", \"excitatory neurons\")){\n",
    "     top_genes <- df[order(-df$avg_log2FC), ][1:20, ]\n",
    "  }\n",
    "\n",
    "  # Save to list\n",
    "  top10_tables[[cell_type]] <- top_genes\n",
    "  \n",
    "  # Optionally print the top genes\n",
    "  print(top_genes)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "8cade43c-7d04-4a2e-be5f-e3494b487081",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "“Scaling data with a low number of groups may produce misleading results”\n",
      "\u001b[1m\u001b[22mScale for \u001b[32mcolour\u001b[39m is already present.\n",
      "Adding another scale for \u001b[32mcolour\u001b[39m, which will replace the existing scale.\n",
      "\u001b[1m\u001b[22mScale for \u001b[32msize\u001b[39m is already present.\n",
      "Adding another scale for \u001b[32msize\u001b[39m, which will replace the existing scale.\n",
      "Warning message:\n",
      "“Scaling data with a low number of groups may produce misleading results”\n",
      "\u001b[1m\u001b[22mScale for \u001b[32mcolour\u001b[39m is already present.\n",
      "Adding another scale for \u001b[32mcolour\u001b[39m, which will replace the existing scale.\n",
      "\u001b[1m\u001b[22mScale for \u001b[32msize\u001b[39m is already present.\n",
      "Adding another scale for \u001b[32msize\u001b[39m, which will replace the existing scale.\n",
      "Warning message:\n",
      "“Scaling data with a low number of groups may produce misleading results”\n",
      "\u001b[1m\u001b[22mScale for \u001b[32mcolour\u001b[39m is already present.\n",
      "Adding another scale for \u001b[32mcolour\u001b[39m, which will replace the existing scale.\n",
      "\u001b[1m\u001b[22mScale for \u001b[32msize\u001b[39m is already present.\n",
      "Adding another scale for \u001b[32msize\u001b[39m, which will replace the existing scale.\n",
      "Warning message:\n",
      "“Scaling data with a low number of groups may produce misleading results”\n",
      "\u001b[1m\u001b[22mScale for \u001b[32mcolour\u001b[39m is already present.\n",
      "Adding another scale for \u001b[32mcolour\u001b[39m, which will replace the existing scale.\n",
      "\u001b[1m\u001b[22mScale for \u001b[32msize\u001b[39m is already present.\n",
      "Adding another scale for \u001b[32msize\u001b[39m, which will replace the existing scale.\n",
      "Warning message:\n",
      "“Scaling data with a low number of groups may produce misleading results”\n",
      "\u001b[1m\u001b[22mScale for \u001b[32mcolour\u001b[39m is already present.\n",
      "Adding another scale for \u001b[32mcolour\u001b[39m, which will replace the existing scale.\n",
      "\u001b[1m\u001b[22mScale for \u001b[32msize\u001b[39m is already present.\n",
      "Adding another scale for \u001b[32msize\u001b[39m, which will replace the existing scale.\n",
      "Warning message:\n",
      "“Scaling data with a low number of groups may produce misleading results”\n",
      "\u001b[1m\u001b[22mScale for \u001b[32mcolour\u001b[39m is already present.\n",
      "Adding another scale for \u001b[32mcolour\u001b[39m, which will replace the existing scale.\n",
      "\u001b[1m\u001b[22mScale for \u001b[32msize\u001b[39m is already present.\n",
      "Adding another scale for \u001b[32msize\u001b[39m, which will replace the existing scale.\n",
      "Warning message:\n",
      "“Scaling data with a low number of groups may produce misleading results”\n",
      "\u001b[1m\u001b[22mScale for \u001b[32mcolour\u001b[39m is already present.\n",
      "Adding another scale for \u001b[32mcolour\u001b[39m, which will replace the existing scale.\n",
      "\u001b[1m\u001b[22mScale for \u001b[32msize\u001b[39m is already present.\n",
      "Adding another scale for \u001b[32msize\u001b[39m, which will replace the existing scale.\n",
      "Warning message:\n",
      "“Scaling data with a low number of groups may produce misleading results”\n",
      "\u001b[1m\u001b[22mScale for \u001b[32mcolour\u001b[39m is already present.\n",
      "Adding another scale for \u001b[32mcolour\u001b[39m, which will replace the existing scale.\n",
      "\u001b[1m\u001b[22mScale for \u001b[32msize\u001b[39m is already present.\n",
      "Adding another scale for \u001b[32msize\u001b[39m, which will replace the existing scale.\n"
     ]
    }
   ],
   "source": [
    "custom_palette <- c('#2166ac','#92c5de','#f7f7f7','#f4a582','#b2182b')\n",
    "\n",
    "# Loop through each cell type in the DEG results\n",
    "for (cell_type in names(deg_results_by_cluster)) {\n",
    "  top_genes <- top10_tables[[cell_type]][1:10,\"gene\"]\n",
    "  if (cell_type %in% c(\"inhibitory neurons\", \"excitatory neurons\")){\n",
    "     top_genes <- top10_tables[[cell_type]][1:20,\"gene\"]\n",
    "  }\n",
    "  \n",
    "  # Subset Seurat object for this cell type\n",
    "  sub_obj <- subset(ifnb_testct, subset = celltype == cell_type)\n",
    "  \n",
    "  # Keep only genes that exist in this object\n",
    "  available_genes <- intersect(top_genes, rownames(sub_obj))\n",
    "  \n",
    "  # Skip if no genes available\n",
    "  if (length(available_genes) == 0) next\n",
    "\n",
    "  # Get plot data first\n",
    "  plot_obj <- DotPlot(\n",
    "    object   = sub_obj,\n",
    "    features = available_genes,\n",
    "    group.by = \"disease\"\n",
    "  )\n",
    "  plot_data <- plot_obj$data\n",
    "  \n",
    "  # Calculate dynamic size breaks for percent expressed\n",
    "  min_pct <- min(plot_data$pct.exp)\n",
    "  max_pct <- max(plot_data$pct.exp)\n",
    "  breaks <- seq(min_pct, max_pct, length.out = 5)\n",
    "  labels <- paste0(round(breaks), \"%\")\n",
    "  \n",
    "  # Now plot with dynamic size scale\n",
    "  p <- plot_obj +\n",
    "    scale_color_gradientn(colors = custom_palette) +\n",
    "    scale_size_area(\n",
    "      max_size = 15,\n",
    "      breaks = breaks,\n",
    "      labels = labels,\n",
    "      name = \"Percent Expressed\"\n",
    "    ) +\n",
    "    ggtitle(paste(\"Cell Type:\", cell_type, \"- Top 10 DEGs for RBD vs HC\")) +\n",
    "    RotatedAxis() +\n",
    "    theme(axis.text.x = element_text(angle = 45, hjust = 1))\n",
    "  \n",
    "  # Save\n",
    "  out_file <- paste0(\"~/Documents/HCPDRBD/man_figs_pdf/deg_dot_plot/RBD_vs_HC_CellType_\", cell_type, \".pdf\")\n",
    "  ggsave(filename = out_file, plot = p, device = \"pdf\", width = 10, height = 5)\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b3f7c31a-cd69-46d1-b344-d5afb53bada3",
   "metadata": {},
   "source": [
    "## GO Enrichment analysis on DEGs for each cell type"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "6f37ffc4-f9d8-4550-8e62-0404e213c30f",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "go_analysis <- function(deg_results_by_cluster, log2FC_threshold = 0.0){\n",
    "    set.seed(384)\n",
    "    # Define the database(s) to use for EnrichR, focusing on GO Biological Process\n",
    "    dbs <- c(\"GO_Biological_Process_2021\")\n",
    "    \n",
    "    # Initialize a list to store enrichment results\n",
    "    enrichment_results <- list()\n",
    "    marker_counts <- list() \n",
    "    \n",
    "    # Loop over each cluster in deg_results_by_cluster to perform enrichment analysis\n",
    "    for (cluster_name in names(deg_results_by_cluster)) {\n",
    "      # Retrieve DEGs for the current cluster from deg_results_by_cluster\n",
    "      cluster_data <- deg_results_by_cluster[[cluster_name]]\n",
    "      \n",
    "      # Select significant DEGs (e.g., adjusted p-value < 0.05 and log fold change threshold)\n",
    "      significant_genes <- rownames(cluster_data[cluster_data$p_val_adj < 0.05 & cluster_data$avg_log2FC > log2FC_threshold, ])\n",
    "      marker_counts[[cluster_name]] <- length(significant_genes)\n",
    "        \n",
    "      # Perform enrichment analysis if there are significant genes\n",
    "      if (length(significant_genes) > 0) {\n",
    "        enrich_result <- enrichr(significant_genes, dbs)\n",
    "        go_results <- enrich_result[[\"GO_Biological_Process_2021\"]]\n",
    "        \n",
    "        # Sort results by adjusted p-value and select the top 10 processes\n",
    "        go_results <- go_results[order(go_results$Adjusted.P.value), ]\n",
    "        top_processes <- head(go_results, 15)\n",
    "    \n",
    "        if (nrow(top_processes) != 0) {\n",
    "            # Initialize list to store the average fold change for each GO term\n",
    "            avg_fc_for_terms <- numeric(nrow(top_processes))\n",
    "            \n",
    "            # Calculate the average fold change for each GO term in top_processes\n",
    "            for (j in 1:nrow(top_processes)) {\n",
    "              if (!is.na(top_processes$Genes[j]) && top_processes$Genes[j] != \"\") {\n",
    "                genes_in_term <- strsplit(as.character(top_processes$Genes[j]), \";\")[[1]]  # Split the genes string by semicolon\n",
    "                \n",
    "                # Only keep genes that are present in cluster_data\n",
    "                genes_in_term <- genes_in_term[genes_in_term %in% rownames(cluster_data)]\n",
    "                  \n",
    "                # Retrieve fold changes for these genes\n",
    "                if (length(genes_in_term) > 0) {\n",
    "                  fc_values <- cluster_data[genes_in_term, \"avg_log2FC\"]\n",
    "                  avg_fc_for_terms[j] <- mean(fc_values, na.rm = TRUE)  # Calculate the average, ignoring NA values\n",
    "                } else {\n",
    "                  avg_fc_for_terms[j] <- NA  # If no genes are found, set to NA\n",
    "                }\n",
    "              } else {\n",
    "                avg_fc_for_terms[j] <- NA  # If 'Genes' is NA, set to NA\n",
    "              }\n",
    "            }\n",
    "        \n",
    "            # Add the average fold changes as a new column in top_processes\n",
    "            top_processes$Avg_Log2FC <- avg_fc_for_terms\n",
    "        }\n",
    "        # Store the top GO processes for this cluster\n",
    "        enrichment_results[[cluster_name]] <- top_processes\n",
    "      }\n",
    "    }\n",
    "    return(list(marker_counts = marker_counts, enrichment_results = enrichment_results))\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "9c680f26-1368-44df-89c5-88b7ec656f9e",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Uploading data to Enrichr... Done.\n",
      "  Querying GO_Biological_Process_2021... Done.\n",
      "Parsing results... Done.\n",
      "Uploading data to Enrichr... Done.\n",
      "  Querying GO_Biological_Process_2021... Done.\n",
      "Parsing results... Done.\n",
      "Uploading data to Enrichr... Done.\n",
      "  Querying GO_Biological_Process_2021... Done.\n",
      "Parsing results... Done.\n",
      "Uploading data to Enrichr... Done.\n",
      "  Querying GO_Biological_Process_2021... Done.\n",
      "Parsing results... Done.\n",
      "Uploading data to Enrichr... Done.\n",
      "  Querying GO_Biological_Process_2021... Done.\n",
      "Parsing results... Done.\n",
      "Uploading data to Enrichr... Done.\n",
      "  Querying GO_Biological_Process_2021... Done.\n",
      "Parsing results... Done.\n",
      "Uploading data to Enrichr... Done.\n",
      "  Querying GO_Biological_Process_2021... Done.\n",
      "Parsing results... Done.\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "<environment: R_GlobalEnv>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "set.seed(384)\n",
    "deg_results_by_cluster_filt <- lapply(deg_results_by_cluster, function(df) {\n",
    "  genes <- rownames(df)\n",
    "  keep  <- (!grepl(\"^(RPL|RPS)\", genes)) & (!grepl(\"^MT-\", genes)) & (!grepl(\"\\\\.\", genes))\n",
    "  df[keep, , drop = FALSE]\n",
    "})\n",
    "output <- go_analysis(deg_results_by_cluster_filt, log2FC_threshold = 0.25)\n",
    "list2env(output, envir = .GlobalEnv)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "8db58f91-ef5b-496b-902d-b794be286bc9",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<dl>\n",
       "\t<dt>$`excitatory neurons`</dt>\n",
       "\t\t<dd>3905</dd>\n",
       "\t<dt>$`oligodendrocyte precursor cells`</dt>\n",
       "\t\t<dd>359</dd>\n",
       "\t<dt>$oligodendrocytes</dt>\n",
       "\t\t<dd>4116</dd>\n",
       "\t<dt>$`inhibitory neurons`</dt>\n",
       "\t\t<dd>777</dd>\n",
       "\t<dt>$astrocytes</dt>\n",
       "\t\t<dd>4440</dd>\n",
       "\t<dt>$microglia</dt>\n",
       "\t\t<dd>1637</dd>\n",
       "\t<dt>$`endothelia cells`</dt>\n",
       "\t\t<dd>165</dd>\n",
       "</dl>\n"
      ],
      "text/latex": [
       "\\begin{description}\n",
       "\\item[\\$`excitatory neurons`] 3905\n",
       "\\item[\\$`oligodendrocyte precursor cells`] 359\n",
       "\\item[\\$oligodendrocytes] 4116\n",
       "\\item[\\$`inhibitory neurons`] 777\n",
       "\\item[\\$astrocytes] 4440\n",
       "\\item[\\$microglia] 1637\n",
       "\\item[\\$`endothelia cells`] 165\n",
       "\\end{description}\n"
      ],
      "text/markdown": [
       "$`excitatory neurons`\n",
       ":   3905\n",
       "$`oligodendrocyte precursor cells`\n",
       ":   359\n",
       "$oligodendrocytes\n",
       ":   4116\n",
       "$`inhibitory neurons`\n",
       ":   777\n",
       "$astrocytes\n",
       ":   4440\n",
       "$microglia\n",
       ":   1637\n",
       "$`endothelia cells`\n",
       ":   165\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "$`excitatory neurons`\n",
       "[1] 3905\n",
       "\n",
       "$`oligodendrocyte precursor cells`\n",
       "[1] 359\n",
       "\n",
       "$oligodendrocytes\n",
       "[1] 4116\n",
       "\n",
       "$`inhibitory neurons`\n",
       "[1] 777\n",
       "\n",
       "$astrocytes\n",
       "[1] 4440\n",
       "\n",
       "$microglia\n",
       "[1] 1637\n",
       "\n",
       "$`endothelia cells`\n",
       "[1] 165\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "marker_counts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1ea75f20-97b9-489a-b8cd-81a13bcaed38",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Logp Plot saved\"\n",
      "[1] \"Log2FC File saved\"\n"
     ]
    }
   ],
   "source": [
    "set.seed(384) \n",
    "# Prepare matrices for plotting\n",
    "cluster_names <- c(\"astrocytes\", \"endothelia cells\", \"excitatory neurons\", \"inhibitory neurons\", \"microglia\", \"oligodendrocytes\", \"oligodendrocyte precursor cells\")\n",
    "# Collect unique biological processes across all clusters\n",
    "all_processes <- unique(unlist(lapply(enrichment_results, function(x) x$Term)))\n",
    "enrichment_matrix <- matrix(\n",
    "  NA, nrow = length(all_processes), ncol = length(cluster_names),\n",
    "  dimnames = list(all_processes, cluster_names)\n",
    ")\n",
    "fc_matrix <- matrix(\n",
    "  NA, nrow = length(all_processes), ncol = length(cluster_names),\n",
    "  dimnames = list(all_processes, cluster_names)\n",
    ")\n",
    "\n",
    "# Fill the matrices with -log10(adjusted p-value) and Avg_Log2FC for each process and cluster\n",
    "for (cluster_name in cluster_names) {\n",
    "  if (!is.null(enrichment_results[[cluster_name]]) && nrow(enrichment_results[[cluster_name]]) > 0) {\n",
    "    for (j in 1:nrow(enrichment_results[[cluster_name]])) {\n",
    "      process <- enrichment_results[[cluster_name]]$Term[j]\n",
    "      pval <- enrichment_results[[cluster_name]]$Adjusted.P.value[j]\n",
    "      avg_log2fc <- enrichment_results[[cluster_name]]$Avg_Log2FC[j]\n",
    "  \n",
    "      # Fill p-value matrix with -log10(p-value)\n",
    "      enrichment_matrix[process, cluster_name] <- -log10(pval)\n",
    "      \n",
    "      # Fill fold change matrix with Avg_Log2FC\n",
    "      fc_matrix[process, cluster_name] <- avg_log2fc\n",
    "      \n",
    "    }\n",
    "  }\n",
    "}   \n",
    "\n",
    "# Replace NA values with a low number (e.g., 0) for better plotting\n",
    "enrichment_matrix[is.na(enrichment_matrix)] <- 0\n",
    "\n",
    "output_pdf = \"../../man_figs_pdf/allct_deg_go_RBDvsHC_logp+.pdf\"\n",
    "# Plot the heatmap using pheatmap\n",
    "pheatmap::pheatmap(enrichment_matrix,\n",
    "                   color = c(\n",
    "                     colorRampPalette(c(\"white\", '#ef3b2c'))(10),  # More resolution for small values\n",
    "                     colorRampPalette(c('#ef3b2c','#cb181d','#a50f15','#67000d'))(90) # High contrast for larger values\n",
    "                   ),\n",
    "                   na_col = \"grey\",           # Use grey for NA values (e.g., processes not enriched in clusters)\n",
    "                   main = \"Enriched BP of Microglial State Signatures (-log10 p-value)\",\n",
    "                   cluster_rows = FALSE,       # Cluster processes for better visualization\n",
    "                   cluster_cols = FALSE,      # Do not cluster clusters (retain the original order)\n",
    "                   angle_col = 45,            # Rotate column names for readability\n",
    "                   fontsize_row = 6,          # Font size for process names\n",
    "                   fontsize_col = 8,\n",
    "                   filename = output_pdf)         # Font size for cluster labels\n",
    "print(\"Logp Plot saved\")\n",
    "\n",
    "# Replace NA values with a low number (e.g., 0) for better plotting\n",
    "fc_matrix[is.na(fc_matrix)] <- 0\n",
    "\n",
    "# Define the custom color palettes\n",
    "# negative_colors <- colorRampPalette(c('#f7fbff','#deebf7','#c6dbef','#9ecae1','#6baed6','#4292c6', '#2171b5', '#084594'))(50)\n",
    "negative_colors <- colorRampPalette(c('#084594','#2171b5','#4292c6','#6baed6','#9ecae1','#c6dbef', '#deebf7', '#f7fbff'))(50)\n",
    "positive_colors <- colorRampPalette(c('#fff5f0','#fee0d2','#fcbba1','#fc9272','#fb6a4a','#ef3b2c','#cb181d','#99000d'))(50)\n",
    "\n",
    "# Combine palettes with white at 0\n",
    "custom_colors <- c(negative_colors, \"white\", positive_colors)\n",
    "\n",
    "# Define custom breaks to center the color scale at 0\n",
    "max_abs_value <- max(abs(fc_matrix))\n",
    "breaks <- seq(-max_abs_value, max_abs_value, length.out = length(custom_colors) + 1)\n",
    "\n",
    "output_pdf = \"../../man_figs_pdf/allct_deg_go_RBDvsHC_log2fc+.pdf\"\n",
    "pheatmap::pheatmap(fc_matrix,\n",
    "                   color = custom_colors,\n",
    "                   breaks = breaks,              # Use custom breaks for diverging colors\n",
    "                   na_col = \"grey\",              # Use grey for NA values\n",
    "                   main = \"Average Log2 Fold Change\",\n",
    "                   cluster_rows = FALSE,          # Cluster processes for better visualization\n",
    "                   cluster_cols = FALSE,          # Cluster clusters for better visualization\n",
    "                   angle_col = 45,               # Rotate column names for readability\n",
    "                   fontsize_row = 6,             # Font size for process names\n",
    "                   fontsize_col = 8,\n",
    "                   filename = output_pdf)             # Font size for cluster labels\n",
    "print(\"Log2FC File saved\")                                                  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "edcbc53d-7892-4659-9e82-652dc31f5de4",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "library(openxlsx)\n",
    "# Create a new workbook\n",
    "wb <- createWorkbook()\n",
    "\n",
    "for (cluster_name in cluster_names) {\n",
    "    # Filter the data for each cluster\n",
    "    cluster_data <- enrichment_results[[cluster_name]]\n",
    "    \n",
    "    # Add a worksheet with the cluster name\n",
    "    addWorksheet(wb, sheetName = cluster_name)\n",
    "    \n",
    "    # Write the filtered data to the worksheet\n",
    "    writeData(wb, sheet = cluster_name, cluster_data, rowNames = TRUE)\n",
    "}\n",
    "\n",
    "# Save the workbook to an Excel file\n",
    "saveWorkbook(wb, file = \"../../DEG_GO_Analysis/GO_RBD_vs_HC/GO_Results_RBD_vs_HC_allct+.xlsx\", overwrite = TRUE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "01db5b10-b95b-4ae2-8372-b4c15cd5d9e7",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "saveRDS(enrichment_results, \"../../DEG_GO_Analysis/GO_RBD_vs_HC/cached_enrichment_results_allct+.rds\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d0f13751-cf89-4485-baad-532a1ca6d0de",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "# DEG analysis (PD vs RBD) for each cell type"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "147e4376-e5a3-4500-af2f-ede14a0e7ca0",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Skipping ct pericytes - does not contain both PD and RBD cells\n",
      "\n"
     ]
    }
   ],
   "source": [
    "set.seed(384)\n",
    "Idents(ifnb_testct) <- \"celltype\"\n",
    "deg_results_by_cluster <- list()\n",
    "\n",
    "\n",
    "for (ct in unique(ifnb_testct$celltype)) {\n",
    "  ct_subset <- subset(ifnb_testct, idents = ct)\n",
    "  Idents(ct_subset) <- \"disease\"\n",
    "  \n",
    "  # Check if both \"PD\" and \"HC\" are present in the subset\n",
    "  if (all(c(\"PD\", \"RBD\") %in% unique(Idents(ct_subset)))) {\n",
    "    # Perform differential expression analysis between \"PD\" and \"HC\" within this cluster\n",
    "    deg_results <- FindMarkers(ct_subset, ident.1 = \"PD\", ident.2 = \"RBD\")\n",
    "    deg_results_by_cluster[[ct]] <- deg_results\n",
    "  } else {\n",
    "    message(paste(\"Skipping ct\", ct, \"- does not contain both PD and RBD cells\"))\n",
    "  }\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c7370650-19b2-4bb7-87b9-d66f519adb8c",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "library(openxlsx)\n",
    "\n",
    "# Create a new workbook\n",
    "wb <- createWorkbook()\n",
    "\n",
    "for (ct in unique(ifnb_testct$celltype)) {\n",
    "    if (ct != \"pericytes\"){\n",
    "        # Filter the data for each cluster\n",
    "        cluster_data <- deg_results_by_cluster[[ct]] %>% filter(p_val_adj < 0.05) %>% arrange(desc(avg_log2FC))\n",
    "        \n",
    "        # Add a worksheet with the cluster name\n",
    "        addWorksheet(wb, sheetName = ct)\n",
    "        \n",
    "        # Write the filtered data to the worksheet\n",
    "        writeData(wb, sheet = ct, cluster_data, rowNames = TRUE)\n",
    "    }\n",
    "}\n",
    "\n",
    "# Save the workbook to an Excel file\n",
    "saveWorkbook(wb, file = \"../../DEG_GO_Analysis/DEGs_PD_vs_RBD/Significant_DEGs_PD_vs_RBD_ct.xlsx\", overwrite = TRUE)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8e6458bb",
   "metadata": {},
   "source": [
    "### Dot plot for top DEGs per cell type"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "c9aadd2e-9517-492e-8792-d482066acb7a",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Top 10/20 genes for cell type: excitatory neurons \n",
      "                 p_val avg_log2FC pct.1 pct.2     p_val_adj     gene\n",
      "XIST      0.000000e+00   5.507481 0.328 0.009  0.000000e+00     XIST\n",
      "TSIX     3.984471e-156   4.571629 0.096 0.012 1.458356e-151     TSIX\n",
      "C5orf17  2.667049e-169   3.554208 0.152 0.042 9.761668e-165  C5orf17\n",
      "HIST1H1D  6.309214e-15   2.650105 0.010 0.001  2.309235e-10 HIST1H1D\n",
      "ASB17     2.815175e-22   2.130123 0.019 0.005  1.030382e-17    ASB17\n",
      "HLA-DMA   1.559506e-53   2.066855 0.077 0.029  5.707947e-49  HLA-DMA\n",
      "MT1G      7.822864e-19   2.058192 0.023 0.008  2.863246e-14     MT1G\n",
      "HIST1H4C 5.515767e-102   1.960780 0.115 0.036  2.018826e-97 HIST1H4C\n",
      "HSPB1     2.331986e-91   1.804235 0.127 0.048  8.535301e-87    HSPB1\n",
      "SERPINE1  5.491860e-09   1.623365 0.012 0.005  2.010076e-04 SERPINE1\n",
      "HIST1H1C  8.500700e-07   1.557737 0.011 0.005  3.111341e-02 HIST1H1C\n",
      "EIF3CL    5.622979e-10   1.535356 0.019 0.009  2.058066e-05   EIF3CL\n",
      "ALLC      8.259291e-17   1.525172 0.025 0.010  3.022983e-12     ALLC\n",
      "PURPL     4.980107e-08   1.486474 0.025 0.015  1.822769e-03    PURPL\n",
      "PDGFD    2.670230e-110   1.481315 0.242 0.128 9.773308e-106    PDGFD\n",
      "NPC1L1    3.172013e-08   1.464130 0.015 0.007  1.160988e-03   NPC1L1\n",
      "RAMAC     1.128372e-19   1.430418 0.029 0.011  4.129955e-15    RAMAC\n",
      "CARTPT    2.124148e-07   1.361568 0.028 0.018  7.774594e-03   CARTPT\n",
      "ERVH-1    5.618928e-07   1.338949 0.011 0.005  2.056584e-02   ERVH-1\n",
      "UBL5      1.197754e-44   1.336335 0.111 0.058  4.383898e-40     UBL5\n",
      "\n",
      "Top 10/20 genes for cell type: oligodendrocyte precursor cells \n",
      "                 p_val avg_log2FC pct.1 pct.2    p_val_adj      gene\n",
      "SPON1-AS1 7.924305e-07   2.395810 0.018 0.003 2.900375e-02 SPON1-AS1\n",
      "ISG15     1.176316e-13   2.310516 0.059 0.017 4.305435e-09     ISG15\n",
      "HSPA1A    2.924118e-26   2.235213 0.175 0.078 1.070257e-21    HSPA1A\n",
      "HSPA1B    8.862370e-19   2.088986 0.113 0.045 3.243716e-14    HSPA1B\n",
      "IFIT1     1.046329e-18   1.777076 0.111 0.043 3.829668e-14     IFIT1\n",
      "PLEKHG4B  3.249211e-10   1.756478 0.036 0.008 1.189244e-05  PLEKHG4B\n",
      "GAL3ST2   8.269080e-07   1.590958 0.024 0.006 3.026566e-02   GAL3ST2\n",
      "APOLD1    7.691688e-29   1.584233 0.156 0.056 2.815235e-24    APOLD1\n",
      "DNAJB1    1.094543e-18   1.520216 0.106 0.039 4.006136e-14    DNAJB1\n",
      "IFI6      7.375297e-13   1.450943 0.101 0.047 2.699432e-08      IFI6\n",
      "\n",
      "Top 10/20 genes for cell type: oligodendrocytes \n",
      "               p_val avg_log2FC pct.1 pct.2     p_val_adj   gene\n",
      "HSPA1A  0.000000e+00   3.904909 0.365 0.075  0.000000e+00 HSPA1A\n",
      "HSPA1B  0.000000e+00   3.662862 0.295 0.046  0.000000e+00 HSPA1B\n",
      "HSPB1   0.000000e+00   2.917165 0.093 0.012  0.000000e+00  HSPB1\n",
      "FERMT3 5.552238e-151   2.872333 0.031 0.003 2.032175e-146 FERMT3\n",
      "BAG3    0.000000e+00   2.698011 0.081 0.011  0.000000e+00   BAG3\n",
      "DNAJB1  0.000000e+00   2.611093 0.178 0.033  0.000000e+00 DNAJB1\n",
      "SLC5A3  0.000000e+00   2.385320 0.206 0.055  0.000000e+00 SLC5A3\n",
      "CKS2    1.473022e-46   2.355623 0.012 0.002  5.391406e-42   CKS2\n",
      "MRPS6   0.000000e+00   2.277005 0.192 0.059  0.000000e+00  MRPS6\n",
      "STIP1   0.000000e+00   2.267912 0.225 0.049  0.000000e+00  STIP1\n",
      "\n",
      "Top 10/20 genes for cell type: inhibitory neurons \n",
      "                 p_val avg_log2FC pct.1 pct.2     p_val_adj     gene\n",
      "XIST      0.000000e+00  4.3615111 0.296 0.013  0.000000e+00     XIST\n",
      "CCK       1.643888e-91  1.2231789 0.286 0.153  6.016795e-87      CCK\n",
      "NAP1L3   4.457956e-118  0.9813438 0.320 0.160 1.631657e-113   NAP1L3\n",
      "CXCL14    4.466589e-70  0.9671926 0.265 0.149  1.634816e-65   CXCL14\n",
      "SCG2      3.472650e-77  0.9501335 0.281 0.158  1.271025e-72     SCG2\n",
      "HSP90AA1 5.143671e-115  0.8753084 0.642 0.509 1.882635e-110 HSP90AA1\n",
      "DNAJA1    2.293668e-78  0.8572217 0.279 0.155  8.395054e-74   DNAJA1\n",
      "RELN      1.482761e-32  0.8495448 0.351 0.267  5.427053e-28     RELN\n",
      "SGO1-AS1  2.574759e-13  0.8154114 0.303 0.261  9.423876e-09 SGO1-AS1\n",
      "NORAD     1.976041e-96  0.7849197 0.370 0.216  7.232508e-92    NORAD\n",
      "HSP90AB1 7.271609e-104  0.7593319 0.539 0.383  2.661482e-99 HSP90AB1\n",
      "NDST4     1.096026e-42  0.7297019 0.262 0.174  4.011564e-38    NDST4\n",
      "HSPH1     1.754817e-80  0.7248578 0.506 0.361  6.422806e-76    HSPH1\n",
      "GAPDH     3.341475e-69  0.7010694 0.279 0.160  1.223013e-64    GAPDH\n",
      "CALM1    5.130831e-112  0.6979266 0.670 0.540 1.877935e-107    CALM1\n",
      "PROX1     2.460534e-62  0.6788501 0.388 0.265  9.005800e-58    PROX1\n",
      "SRGAP2B   8.348822e-73  0.6778859 0.382 0.245  3.055752e-68  SRGAP2B\n",
      "NXPH2     1.101564e-44  0.6506965 0.343 0.245  4.031834e-40    NXPH2\n",
      "VAMP2     1.813647e-70  0.6389082 0.343 0.213  6.638128e-66    VAMP2\n",
      "PWRN4     7.217496e-92  0.6388828 0.500 0.335  2.641676e-87    PWRN4\n",
      "\n",
      "Top 10/20 genes for cell type: astrocytes \n",
      "                 p_val avg_log2FC pct.1 pct.2     p_val_adj     gene\n",
      "SH3BGR   2.461911e-184   2.182471 0.307 0.127 9.010840e-180   SH3BGR\n",
      "HSPA1A   2.651616e-172   1.793183 0.331 0.155 9.705180e-168   HSPA1A\n",
      "HSPH1    1.895253e-107   1.772517 0.395 0.267 6.936815e-103    HSPH1\n",
      "KITLG    2.204840e-226   1.699284 0.327 0.117 8.069936e-222    KITLG\n",
      "CRYAB    6.290010e-187   1.677109 0.385 0.191 2.302207e-182    CRYAB\n",
      "XIST     5.036257e-300   1.552854 0.330 0.078 1.843320e-295     XIST\n",
      "HSP90AA1 2.125209e-124   1.520073 0.629 0.497 7.778477e-120 HSP90AA1\n",
      "HSPA1B    3.792844e-95   1.423142 0.252 0.131  1.388219e-90   HSPA1B\n",
      "DIAPH3   2.599267e-166   1.349246 0.272 0.100 9.513576e-162   DIAPH3\n",
      "TXNIP    4.103663e-146   1.339198 0.277 0.114 1.501982e-141    TXNIP\n",
      "\n",
      "Top 10/20 genes for cell type: microglia \n",
      "                   p_val avg_log2FC pct.1 pct.2     p_val_adj       gene\n",
      "HSPH1       3.846516e-29  1.9101963 0.360 0.250  1.407863e-24      HSPH1\n",
      "IFI44L      6.231779e-25  1.3353210 0.304 0.201  2.280893e-20     IFI44L\n",
      "XIST        6.159309e-79  1.3234358 0.307 0.095  2.254369e-74       XIST\n",
      "DLEU7       3.434797e-43  1.2469349 0.344 0.193  1.257170e-38      DLEU7\n",
      "CPED1      1.760133e-130  1.2309461 0.746 0.489 6.442262e-126      CPED1\n",
      "CSGALNACT1 5.746732e-157  1.1664656 0.798 0.509 2.103362e-152 CSGALNACT1\n",
      "HSP90AA1    9.250438e-09  1.0368626 0.430 0.387  3.385753e-04   HSP90AA1\n",
      "FOXP2      6.183576e-102  1.0330535 0.774 0.586  2.263251e-97      FOXP2\n",
      "ITM2B      2.697874e-103  1.0083027 0.752 0.542  9.874488e-99      ITM2B\n",
      "ABCA1       3.534874e-27  0.9924352 0.262 0.151  1.293799e-22      ABCA1\n",
      "\n",
      "Top 10/20 genes for cell type: endothelia cells \n",
      "                   p_val avg_log2FC pct.1 pct.2    p_val_adj        gene\n",
      "PAUPAR      1.026509e-07   3.701401 0.054 0.007 3.757126e-03      PAUPAR\n",
      "CFAP47      1.536838e-07   3.544653 0.070 0.015 5.624979e-03      CFAP47\n",
      "MIR4300HG   2.715067e-11   3.188060 0.127 0.034 9.937418e-07   MIR4300HG\n",
      "IFIT1       1.048361e-18   3.004187 0.197 0.047 3.837105e-14       IFIT1\n",
      "EMID1       2.535185e-07   2.984821 0.071 0.016 9.279032e-03       EMID1\n",
      "OBI1-AS1    2.289680e-11   2.734058 0.166 0.059 8.380458e-07    OBI1-AS1\n",
      "IFIT2       4.703121e-10   2.705388 0.106 0.026 1.721389e-05       IFIT2\n",
      "FAM198B-AS1 2.215191e-08   2.699438 0.088 0.022 8.107822e-04 FAM198B-AS1\n",
      "HSPA1A      5.773899e-11   2.679432 0.330 0.198 2.113305e-06      HSPA1A\n",
      "RP1         2.996627e-09   2.650796 0.123 0.039 1.096796e-04         RP1\n"
     ]
    }
   ],
   "source": [
    "top10_tables <- list()\n",
    "\n",
    "for (cell_type in names(deg_results_by_cluster)) {\n",
    "  cat(\"\\nTop 10/20 genes for cell type:\", cell_type, \"\\n\")\n",
    "  \n",
    "  # Get the dataframe\n",
    "  df <- deg_results_by_cluster[[cell_type]] %>% filter(p_val_adj < 0.05)\n",
    "    \n",
    "  # Add gene names as a column if they're rownames\n",
    "  df$gene <- rownames(df)\n",
    "    \n",
    "  df <- df[!grepl(\"^(LINC)\", df$gene), ]\n",
    "  df <- df[!grepl(\"^(NPIPA8|TTTY)\", df$gene), ]\n",
    "  df <- df[!grepl(\"^(MT-)\", df$gene), ]\n",
    "  df <- df[!grepl(\"\\\\.\\\\d+$\", df$gene), ]\n",
    "    \n",
    "  if (cell_type %in% c(\"microglia\", \"inhibitory neurons\", \"astrocytes\")){\n",
    "      df <- df %>% filter(pct.1 >= 0.25 | pct.2 >= 0.25)\n",
    "  }\n",
    "  \n",
    "  # Select top 10 genes with highest avg_log2FC\n",
    "  top_genes <- df[order(-df$avg_log2FC), ][1:10, ]\n",
    "\n",
    "  if (cell_type %in% c(\"inhibitory neurons\", \"excitatory neurons\")){\n",
    "     top_genes <- df[order(-df$avg_log2FC), ][1:20, ]\n",
    "  }\n",
    "\n",
    "  # Save to list\n",
    "  top10_tables[[cell_type]] <- top_genes\n",
    "  \n",
    "  # Optionally print the top genes\n",
    "  print(top_genes)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "452a2714-84e7-4e6a-a0e6-558355390b90",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "“Scaling data with a low number of groups may produce misleading results”\n",
      "\u001b[1m\u001b[22mScale for \u001b[32mcolour\u001b[39m is already present.\n",
      "Adding another scale for \u001b[32mcolour\u001b[39m, which will replace the existing scale.\n",
      "\u001b[1m\u001b[22mScale for \u001b[32msize\u001b[39m is already present.\n",
      "Adding another scale for \u001b[32msize\u001b[39m, which will replace the existing scale.\n",
      "Warning message:\n",
      "“Scaling data with a low number of groups may produce misleading results”\n",
      "\u001b[1m\u001b[22mScale for \u001b[32mcolour\u001b[39m is already present.\n",
      "Adding another scale for \u001b[32mcolour\u001b[39m, which will replace the existing scale.\n",
      "\u001b[1m\u001b[22mScale for \u001b[32msize\u001b[39m is already present.\n",
      "Adding another scale for \u001b[32msize\u001b[39m, which will replace the existing scale.\n",
      "Warning message:\n",
      "“Scaling data with a low number of groups may produce misleading results”\n",
      "\u001b[1m\u001b[22mScale for \u001b[32mcolour\u001b[39m is already present.\n",
      "Adding another scale for \u001b[32mcolour\u001b[39m, which will replace the existing scale.\n",
      "\u001b[1m\u001b[22mScale for \u001b[32msize\u001b[39m is already present.\n",
      "Adding another scale for \u001b[32msize\u001b[39m, which will replace the existing scale.\n",
      "Warning message:\n",
      "“Scaling data with a low number of groups may produce misleading results”\n",
      "\u001b[1m\u001b[22mScale for \u001b[32mcolour\u001b[39m is already present.\n",
      "Adding another scale for \u001b[32mcolour\u001b[39m, which will replace the existing scale.\n",
      "\u001b[1m\u001b[22mScale for \u001b[32msize\u001b[39m is already present.\n",
      "Adding another scale for \u001b[32msize\u001b[39m, which will replace the existing scale.\n",
      "Warning message:\n",
      "“Scaling data with a low number of groups may produce misleading results”\n",
      "\u001b[1m\u001b[22mScale for \u001b[32mcolour\u001b[39m is already present.\n",
      "Adding another scale for \u001b[32mcolour\u001b[39m, which will replace the existing scale.\n",
      "\u001b[1m\u001b[22mScale for \u001b[32msize\u001b[39m is already present.\n",
      "Adding another scale for \u001b[32msize\u001b[39m, which will replace the existing scale.\n",
      "Warning message:\n",
      "“Scaling data with a low number of groups may produce misleading results”\n",
      "\u001b[1m\u001b[22mScale for \u001b[32mcolour\u001b[39m is already present.\n",
      "Adding another scale for \u001b[32mcolour\u001b[39m, which will replace the existing scale.\n",
      "\u001b[1m\u001b[22mScale for \u001b[32msize\u001b[39m is already present.\n",
      "Adding another scale for \u001b[32msize\u001b[39m, which will replace the existing scale.\n",
      "Warning message:\n",
      "“Scaling data with a low number of groups may produce misleading results”\n",
      "\u001b[1m\u001b[22mScale for \u001b[32mcolour\u001b[39m is already present.\n",
      "Adding another scale for \u001b[32mcolour\u001b[39m, which will replace the existing scale.\n",
      "\u001b[1m\u001b[22mScale for \u001b[32msize\u001b[39m is already present.\n",
      "Adding another scale for \u001b[32msize\u001b[39m, which will replace the existing scale.\n"
     ]
    }
   ],
   "source": [
    "custom_palette <- c('#2166ac','#92c5de','#f7f7f7','#f4a582','#b2182b')\n",
    "\n",
    "# Loop through each cell type in the DEG results\n",
    "for (cell_type in names(deg_results_by_cluster)) {\n",
    "  top_genes <- top10_tables[[cell_type]][1:10,\"gene\"]\n",
    "  if (cell_type %in% c(\"inhibitory neurons\", \"excitatory neurons\")){\n",
    "     top_genes <- top10_tables[[cell_type]][1:20,\"gene\"]\n",
    "  }\n",
    "  \n",
    "  # Subset Seurat object for this cell type\n",
    "  sub_obj <- subset(ifnb_testct, subset = celltype == cell_type)\n",
    "  \n",
    "  # Keep only genes that exist in this object\n",
    "  available_genes <- intersect(top_genes, rownames(sub_obj))\n",
    "  \n",
    "  # Skip if no genes available\n",
    "  if (length(available_genes) == 0) next\n",
    "\n",
    "  # Get plot data first\n",
    "  plot_obj <- DotPlot(\n",
    "    object   = sub_obj,\n",
    "    features = available_genes,\n",
    "    group.by = \"disease\"\n",
    "  )\n",
    "  plot_data <- plot_obj$data\n",
    "  \n",
    "  # Calculate dynamic size breaks for percent expressed\n",
    "  min_pct <- min(plot_data$pct.exp)\n",
    "  max_pct <- max(plot_data$pct.exp)\n",
    "  breaks <- seq(min_pct, max_pct, length.out = 5)\n",
    "  labels <- paste0(round(breaks), \"%\")\n",
    "  \n",
    "  # Now plot with dynamic size scale\n",
    "  p <- plot_obj +\n",
    "    scale_color_gradientn(colors = custom_palette) +\n",
    "    scale_size_area(\n",
    "      max_size = 15,\n",
    "      breaks = breaks,\n",
    "      labels = labels,\n",
    "      name = \"Percent Expressed\"\n",
    "    ) +\n",
    "    ggtitle(paste(\"Cell Type:\", cell_type, \"- Top 10 DEGs for PD vs RBD\")) +\n",
    "    RotatedAxis() +\n",
    "    theme(axis.text.x = element_text(angle = 45, hjust = 1))\n",
    "  \n",
    "  # Save\n",
    "  out_file <- paste0(\"~/Documents/HCPDRBD/man_figs_pdf/deg_dot_plot/PD_vs_RBD_CellType_\", cell_type, \".pdf\")\n",
    "  ggsave(filename = out_file, plot = p, device = \"pdf\", width = 10, height = 5)\n",
    "}"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.3.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
